// @webex/internal-plugin-ediscovery/src/ediscovery.js
/* eslint-disable prefer-template */
/*!
 * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
 */
import {waitForValue, WebexPlugin} from '@webex/webex-core';
import {oneFlight} from '@webex/common';
import Map from 'babel-runtime/core-js/map';

import {InvalidEmailAddressError} from './ediscovery-error';

const defaultOptions = {timeoutMs: 30000}; // 30s default timeout

/**
 * @class EDiscovery is used by compliance officers to run compliance reports
 *
 */
const EDiscovery = WebexPlugin.extend({
  namespace: 'EDiscovery',

  session: {
    contentSummaryBySpaceIdCache: {
      type: 'object',
      default() {
        return new Map();
      }
    }
  },

  @waitForValue('@')
  /**
   * Creates a compliance report with a specific set of search parameters
   * @param {Object} reportRequest - A set of criteria for determining the focus of the search
   * @param {Object} options - optional parameters for this method
   * @param {number} options.timeoutMs - connection timeout in milliseconds, defaults to 10s
   * @returns {Promise<ResponseEntity>} Http response containing the new report record
   */
  createReport(reportRequest, options) {
    if (!reportRequest) {
      throw Error('Undefined parameter');
    }

    // use spread operator to set default options
    options = {...defaultOptions, ...options};

    const body = reportRequest;

    return this.request({
      method: 'POST',
      service: 'ediscovery',
      resource: 'reports',
      timeout: options.timeoutMs,
      body
    })
      .catch((reason) => this._handleReportRequestError(reason));
  },

  /**
   * Checks the error from createReport and ensures the appropriate error is sent to the client
   * @param {Error} reason - Error response thrown by the request to createReport
   * @returns {Promise<Error>} Promise rejection containing the error
   */
  _handleReportRequestError(reason) {
    if (reason.body.errorCode === InvalidEmailAddressError.getErrorCode()) {
      try {
        const invalidEmails = JSON.parse(reason.body.message);

        if (Array.isArray(invalidEmails) && invalidEmails.length) {
          const invalidEmailAddressError = new InvalidEmailAddressError(invalidEmails);

          return Promise.reject(invalidEmailAddressError);
        }

        this.logger.warn('InvalidEmailAddress error received but the list could not be parsed to the correct format.');
      }
      catch (error) {
        // assume syntax error and continue
        this.logger.error('InvalidEmailAddress error received but an error occured while parsing the emails.');
      }
    }

    return Promise.reject(reason);
  },

  @waitForValue('@')
  /**
   * Retrieves information relating to a specified report
   * @param {UUID} reportId - Id of the report being requested
   * @param {Object} options - optional parameters for this method
   * @param {number} options.timeoutMs - connection timeout in milliseconds, defaults to 10s
   * @returns {Promise<ResponseEntity<ReportRecord>>} Http response containing the specified report record
   */
  getReport(reportId, options) {
    if (!reportId) {
      throw Error('Undefined parameter');
    }

    // use spread operator to set default options
    options = {...defaultOptions, ...options};

    return this.request({
      method: 'GET',
      service: 'ediscovery',
      resource: `reports/${reportId}`,
      timeout: options.timeoutMs
    });
  },

  @waitForValue('@')
  /**
   * Retrieves all the compliance officers reports
   * @param {Object} options - optional parameters for this method
   * @param {number} options.offset - start position from which to retrieve records
   * @param {number} options.size - the number of records to retrieve
   * @param {number} options.timeoutMs - connection timeout in milliseconds, defaults to 10s
   * @returns {Promise<ResponseEntity<Array<ReportRecord>>>} Http Response containing a list of report records
   */
  getReports(options) {
    // use spread operator to set default options
    options = {...defaultOptions, ...options};

    return this.request({
      method: 'GET',
      service: 'ediscovery',
      resource: 'reports',
      qs: {offset: options.offset, size: options.size},
      timeout: options.timeoutMs
    });
  },

  @waitForValue('@')
  /**
   * Deletes a specified report
   * @param {UUID} reportId - Id of the report being requested for deletion
   * @param {Object} options - optional parameters for this method
   * @param {number} options.timeoutMs - connection timeout in milliseconds, defaults to 10s
   * @returns {Promise<ResponseEntity>} HttpResponse indicating if delete was successful
   */
  deleteReport(reportId, options) {
    if (!reportId) {
      throw Error('Undefined parameter');
    }

    // use spread operator to set default options
    options = {...defaultOptions, ...options};

    return this.request({
      method: 'DELETE',
      service: 'ediscovery',
      resource: `reports/${reportId}`,
      timeout: options.timeoutMs
    });
  },

  @waitForValue('@')
  /**
   * Restarts a completed or cancelled report so that it begins again from scratch
   * @param {UUID} reportId - Id of the report being requested
   * @param {Object} options - optional parameters for this method
   * @param {number} options.timeoutMs - connection timeout in milliseconds, defaults to 10s
   * @returns {Promise<ResponseEntity<ReportRecord>>} Http response containing the report record
   */
  restartReport(reportId, options) {
    if (!reportId) {
      throw Error('Undefined parameter');
    }

    // use spread operator to set default options
    options = {...defaultOptions, ...options};

    return this.request({
      method: 'PUT',
      service: 'ediscovery',
      resource: `reports/${reportId}`,
      timeout: options.timeoutMs
    });
  },

  @waitForValue('@')
  /**
   * Retrieves content associated with a report
   * @param {UUID} reportId - Id of the report which contains the content
   * @param {Object} options - optional parameters for this method
   * @param {number} options.offset - start position from which to retrieve records
   * @param {number} options.size - the number of records to retrieve
   * @param {number} options.timeoutMs - connection timeout in milliseconds, defaults to 10s
   * @returns {Promise<ResponseEntity<[Activity]>>} Http response containing the activities
   */
  getContent(reportId, options) {
    if (!reportId) {
      throw Error('Undefined parameter');
    }

    // use spread operator to set default options
    options = {...defaultOptions, ...options};

    return this.request({
      method: 'GET',
      service: 'ediscovery',
      resource: `reports/${reportId}/contents`,
      qs: {offset: options.offset, size: options.size},
      timeout: options.timeoutMs
    });
  },

  @waitForValue('@')
  @oneFlight({keyFactory: (reportId) => `${reportId}`})
  /**
   * Retrieves a list of the conversations relevant to a specified report
   * @param {UUID} reportId - Id of the report being requested
   * @param {Object} options - optional parameters for this method
   * @param {number} options.timeoutMs - connection timeout in milliseconds, defaults to 90s
   * @returns {Promise<ResponseEntity<ContentSummary>>} Http response containing the content summary
   */
  getContentSummary(reportId, options) {
    if (!reportId) {
      throw Error('Undefined parameter');
    }

    const contentSummaryDefaultOptions = {timeoutMs: 90000};

    // use spread operator to set default options
    options = {...contentSummaryDefaultOptions, ...options};

    let offset = 0;
    const size = 100;
    let contentSummary = [];

    function getNextContentSummaryPage(plugin) {
      return plugin.request({
        method: 'GET',
        service: 'ediscovery',
        resource: `reports/${reportId}/contents/summary`,
        qs: {offset, size},
        timeout: options.timeoutMs
      })
        .then((res) => {
          // The results are written to the contentSummaryCache as information for a space is likely to be accessed very frequently for decryption
          if (!plugin.contentSummaryBySpaceIdCache.has(reportId)) {
            plugin.contentSummaryBySpaceIdCache.set(reportId, new Map());
          }

          for (let i = 0; i < res.body.length; i += 1) {
            if (res.body[i] && res.body[i].spaceId) {
              try {
                plugin.contentSummaryBySpaceIdCache.get(reportId).set(res.body[i].spaceId, res.body[i]);
              }
              catch (error) {
                plugin.logger.error(`Error adding ${res.body[i].spaceId} to contentSummaryBySpaceIdCache: ${error}`);
              }
            }
            else {
              plugin.logger.error('Error adding undefined space to contentSummaryBySpaceIdCache');
            }
          }

          contentSummary = contentSummary.concat(res.body);
          if (res.body.length < size) {
            return contentSummary;
          }
          offset += size;

          return getNextContentSummaryPage(plugin);
        });
    }

    return getNextContentSummaryPage(this);
  },

  @waitForValue('@')
  @oneFlight({keyFactory: (reportId, spaceId) => `${reportId}-${spaceId}`})
  /**
   * Retrieves information for a specific conversation relevant to a specified report
   * @param {UUID} reportId - Id of the report which contains the relevant content summary
   * @param {UUID} spaceId - Id of the conversation being requested
   * @param {Object} options - optional parameters for this method
   * @param {number} options.timeoutMs - connection timeout in milliseconds, defaults to 10s
   * @returns {Promise<ResponseEntity<ContentSummary>>} Http response containing the specified content summary
   */
  getContentSummaryBySpaceId(reportId, spaceId, options) {
    if (!reportId || !spaceId) {
      throw Error('Undefined parameter(s)');
    }

    // use spread operator to set default options
    options = {...defaultOptions, ...options};

    // If the content summary for this space has already been cached then it can be retrieved from there instead of making a network call to ediscovery
    if (this.contentSummaryBySpaceIdCache.has(reportId) && this.contentSummaryBySpaceIdCache.get(reportId).has(spaceId)) {
      return this.contentSummaryBySpaceIdCache.get(reportId).get(spaceId);
    }

    this.logger.warn(`Cache miss for space ${spaceId} in getContentSummaryBySpaceId`);

    return this.request({
      method: 'GET',
      service: 'ediscovery',
      resource: `reports/${reportId}/contents/summary/${spaceId}`,
      timeout: options.timeoutMs
    })
      .then((res) => {
        if (!this.contentSummaryBySpaceIdCache.has(reportId)) {
          this.contentSummaryBySpaceIdCache.set(reportId, new Map());
        }

        if (res.body && res.body.spaceId) {
          try {
            this.contentSummaryBySpaceIdCache.get(reportId).set(res.body.spaceId, res.body);
          }
          catch (error) {
            this.logger.error(`Error adding ${res.body.spaceId} to contentSummaryBySpaceIdCache: ${error}`);
          }
        }
        else {
          this.logger.error('Error adding undefined space to contentSummaryBySpaceIdCache');
        }

        return res.body;
      });
  }

});

export default EDiscovery;
