// @webex/internal-plugin-ediscovery/src/report-request.js

const defaultRange = {
  startTime: {
    year: 2000,
    month: 'JANUARY',
    dayOfMonth: 1,
    dayOfWeek: 'FRIDAY',
    dayOfYear: 1,
    monthValue: 1,
    hour: 1,
    minute: 1,
    second: 0,
    nano: 0,
    chronology: {id: 'ISO', calendarType: 'iso8601'}
  },
  endTime: {
    year: 2000,
    month: 'JANUARY',
    dayOfMonth: 1,
    dayOfWeek: 'FRIDAY',
    dayOfYear: 1,
    monthValue: 1,
    hour: 1,
    minute: 1,
    second: 0,
    nano: 0,
    chronology: {id: 'ISO', calendarType: 'iso8601'}
  }
};

/**
 * Creates a report request object with a specific set of search parameters
 * @param {String} name - A label to identify the report
 * @param {String} description - A textual summary of the reports purpose
 * @param {Array<String>} emails - A list of user emails relevant to the report
 * @param {Array<String>} userIds - A list of UUIDs relevant to the report
 * @param {Array<String>} keywords - A list of search terms relevant to the report
 * @param {Array<String>} spaceNames - A list of space names relevant to the report
 * @param {Array<String>} spaceIds - A list of space ids relevant to the report
 * @param {Object} range - Contains the start time and end time defining the search period
 * @returns {Object} ReportRequest - Contains all search parameters
 */
class ReportRequest {
  constructor(name = '', description = '', emails = [], userIds = [], keywords = [], encryptionKeyUrl = '', spaceNames = [], spaceIds = [], range = defaultRange) {
    this.name = name;
    this.description = description;
    this.emails = emails;
    this.userIds = userIds;
    this.keywords = keywords;
    this.encryptionKeyUrl = encryptionKeyUrl;
    this.spaceNames = spaceNames;
    this.spaceIds = spaceIds;
    this.range = range;
  }
}

export default ReportRequest;
