diff --git a/node_modules/documentation/src/parsers/parse_to_ast.js b/node_modules/documentation/src/parsers/parse_to_ast.js
index 9f5673d..24c6094 100644
--- a/node_modules/documentation/src/parsers/parse_to_ast.js
+++ b/node_modules/documentation/src/parsers/parse_to_ast.js
@@ -27,7 +27,7 @@ function getParserOpts(file) {
     sourceType: 'module',
     plugins: [
       ...standardBabelParserPlugins,
-      ['decorators', { decoratorsBeforeExport: false }],
+      'decorators-legacy',
       ...(TYPESCRIPT_EXTS[path.extname(file || '')] || ['flow', 'jsx'])
     ]
   };
