// @ciscospark/test-helper-mocha/src/index.js
/*!
 * Copyright (c) 2015-2017 Cisco Systems, Inc. See LICENSE file.
 */


const browser = require('bowser');

/**
 * Indicates if we're running in node
 * @returns {boolean}
 * @private
 */
function inNode() {
  return typeof window === 'undefined';
}


/**
 * Indeicates if we're running in a web browser
 * @returns {boolean}
 * @private
 */
function inBrowser() {
  return !inNode();
}

/**
 * Indeicates if we're running in firefox
 * @returns {boolean}
 * @private
 */
function inFirefox() {
  return browser.firefox;
}

/**
 * Indeicates if we're running in internet explorer
 * @returns {boolean}
 * @private
 */
function inIE() {
  return browser.ie;
}

/**
 * noop
 * @returns {undefined}
 * @private
 */
function noop() {
  // intentionally empty
}

module.exports = {
  /**
   * Wrap the desired mochaMethod with `flaky` to indicate it's a flaky test and
   * exclude it from the cloudapps pipeline
   * example:
   * `flaky(it)('sometimes passes and sometimes failes')`
   * @param {Function} mochaMethod `it` or `describe`
   * @returns {Function} mochaMethod or mochaMethod.skip
   */
  flaky: function flaky(mochaMethod) {
    // If mochaMethod doesn't have a skip method, assume that mochaMethod is
    // already either a .skip or a .only
    if (!mochaMethod.skip) {
      return mochaMethod;
    }
    return process.env.SKIP_FLAKY_TESTS ? mochaMethod.skip : mochaMethod;
  },

  inBrowser,

  inNode,

  /**
   * Wrap the desired mochaMethod with `skipInBrowser` to prevent the
   * corresponding test or group of tests from running in a web browser
   * example:
   * `skipInBrowser(it)('does a thing that only works in node')`
   * @param {Function} mochaMethod `it` or `describe`
   * @returns {Function} mochaMethod or mochaMethod.skip
   */
  skipInBrowser: function skipInBrowser(mochaMethod) {
    // If mochaMethod doesn't have a skip method, assume that mochaMethod is
    // already either a .skip or a .only
    if (!mochaMethod.skip) {
      return mochaMethod;
    }
    return inBrowser() ? mochaMethod.skip : mochaMethod;
  },

  /**
   * Wrap the desired mochaMethod with `skipInNode` to prevent the
   * corresponding test or group of tests from running in node.
   * example:
   * `skipInNode(it)('does a thing that only works in a web browser')`
   * @param {Function} mochaMethod `it` or `describe`
   * @returns {Function} mochaMethod or mochaMethod.skip
   */
  skipInNode: function skipInNode(mochaMethod) {
    // If mochaMethod doesn't have a skip method, assume that mochaMethod is
    // already either a .skip or a .only
    if (!mochaMethod.skip) {
      return mochaMethod;
    }
    return inNode() ? mochaMethod.skip : mochaMethod;
  },

  skipInFirefox: function skipInFirefox(mochaMethod) {
    // If mochaMethod doesn't have a skip method, assume that mochaMethod is
    // already either a .skip or a .only
    if (!mochaMethod.skip) {
      return mochaMethod;
    }
    return inFirefox() ? mochaMethod.skip : mochaMethod;
  },

  skipInIE: function skipInIE(mochaMethod) {
    // If mochaMethod doesn't have a skip method, assume that mochaMethod is
    // already either a .skip or a .only
    if (!mochaMethod.skip) {
      return mochaMethod;
    }
    return inIE() ? mochaMethod.skip : mochaMethod;
  },

  /**
   * Similar to skipInNode in that it prevents the test from running, but goes a
   * step further to hide it from the list of skipped tests. Should be used when
   * the test is only valid in webkit browsers
   * @param {Function} mochaMethod
   * @returns {Function}
   */
  webkitOnly: function webkitOnly(mochaMethod) {
    return inWebkit() ? mochaMethod : noop;
  },

  /**
   * Similar to skipInNode in that it prevents the test from running, but goes a
   * step further to hide it from the list of skipped tests. Should be used when
   * the test will never be valid in NodeJS
   * @param {Function} mochaMethod
   * @returns {Function}
   */
  browserOnly: function browserOnly(mochaMethod) {
    return inBrowser() ? mochaMethod : noop;
  },

  /**
   * Similar to skipInBrowser in that it prevents the test from running, but
   * goes a step further to hide it from the list of skipped tests. Should be
   * used when the test will never be valid in a Browser
   * @param {Function} mochaMethod
   * @returns {Function}
   */
  nodeOnly: function nodeOnly(mochaMethod) {
    return inNode() ? mochaMethod : noop;
  },

  /**
   * Returns a promise that resolves when event is fired count times or rejects
   * when max expires
   * @param {number} max
   * @param {number} count
   * @param {string} event
   * @param {EventEmitter} emitter
   * @returns {Promise}
   */
  expectNEvents: function expectNEvents(max, count, event, emitter) {
    let timer;

    let currentCount = 0;

    return Promise.race([
      new Promise((resolve, reject) => {
        timer = setTimeout(() => {
          reject(new Error(`${event} fired only ${currentCount} times within ${max}ms. expected ${count} times`));
        }, max);
      }),
      new Promise((resolve) => {
        emitter.on(event, fn);

        /**
         * helper
         * @private
         * @returns {undefined}
         */
        function fn() {
          currentCount += 1;
          if (currentCount === count) {
            emitter.off(event, fn);
            clearTimeout(timer);
            resolve();
          }
        }
      })
    ]);
  },

  /**
   * @param {number} max
   * @param {string} event
   * @param {EventEmitter} emitter
   * @param {string} msg
   * @returns {Promise<mixed>} Resolves with the results of the event
   */
  expectEvent: function expectEvent(max, event, emitter, msg) {
    let timer;
    return Promise.race([
      new Promise((resolve, reject) => {
        timer = setTimeout(() => {
          reject(new Error(`${event} did not fire within ${max}ms${msg ? `: ${msg}` : ''}`));
        }, max);
      }),
      new Promise((resolve) => {
        emitter.once(event, (result) => {
          clearTimeout(timer);
          resolve(result);
        });
      })
    ]);
  },

  maxWaitForEvent: function maxWaitForEvent(max, event, emitter) {
    return Promise.race([
      new Promise((resolve) => {
        setTimeout(resolve, max);
      }),
      new Promise((resolve) => {
        emitter.once(event, resolve);
      })
    ]);
  },

  maxWaitForPromise: function maxWaitForPromise(timeout, promise) {
    return Promise.race([
      promise,
      new Promise((resolve, reject) => {
        setTimeout(() => {
          reject(new Error(`Timeout of ${timeout} expired before promise completed`));
        }, timeout);
      })
    ]);
  }
};
