// @webex/internal-plugin-conversation/src/threading.js
import {sortBy} from 'lodash';

export const minBatchSize = 10;
export const fetchLoopCountMax = 100;
export const batchSizeIncrementCount = 10;

export const OLDER = 'older';
export const NEWER = 'newer';
export const MID = 'mid';
export const INITIAL = 'initial';

export const ACTIVITY_TYPES = {
  REPLY: 'REPLY',
  EDIT: 'EDIT',
  REACTION: 'REACTION',
  REACTION_SELF: 'REACTION_SELF',
  ROOT: 'ROOT',
  CREATE: 'CREATE',
  TOMBSTONE: 'TOMBSTONE',
  DELETE: 'DELETE'
};

// TODO: create type for tombstone edit? could make it easy to ignore older edits...
export const getActivityType = (activity) => {
  if (activity.activityType === 'reply') {
    return ACTIVITY_TYPES.REPLY;
  }
  if (activity.parent && activity.parent.type === 'edit') {
    return ACTIVITY_TYPES.EDIT;
  }
  if (activity.verb === 'add' || activity.type === 'reactionSummary') {
    return ACTIVITY_TYPES.REACTION;
  }
  if (activity.type === 'reactionSelfSummary') {
    return ACTIVITY_TYPES.REACTION_SELF;
  }
  if (activity.verb === 'create') {
    return ACTIVITY_TYPES.CREATE;
  }
  if (activity.verb === 'tombstone') {
    return ACTIVITY_TYPES.TOMBSTONE;
  }
  if (activity.verb === 'delete') {
    return ACTIVITY_TYPES.DELETE;
  }

  return ACTIVITY_TYPES.ROOT;
};

export const getPublishedDate = (activity = {}) => new Date(activity.published).getTime();

/**
 * @param {Object} activity1
 * @param {Object} activity2
 * @returns {boolean} true if first activity is newer than second
 */
export const isNewer = (activity1, activity2) => getPublishedDate(activity1) > getPublishedDate(activity2);

export const getActivityObjectsFromMap = (hashMap) => Array.from(hashMap).map(([, activity]) => activity);

export const sortActivitiesByPublishedDate = (activities) => sortBy(activities, (activity) => getPublishedDate(activity));

export const getParentId = (activity) => activity && activity.parent && activity.parent.id;

export const isRootActivity = (act) => getActivityType(act) === ACTIVITY_TYPES.ROOT;
export const isReplyActivity = (act) => getActivityType(act) === ACTIVITY_TYPES.REPLY;
export const isEditActivity = (act) => getActivityType(act) === ACTIVITY_TYPES.EDIT;
export const isCreateActivity = (act) => getActivityType(act) === ACTIVITY_TYPES.CREATE;
export const isDeleteActivity = (act) => getActivityType(act) === ACTIVITY_TYPES.DELETE;

// use accessors for ease of refactoring underlying implementation
/**
       * @param {Map} destination destination map object for data. Currently a Map, but could be refactored to use any type.
       * @param {string} key
       * @param {any} value
       * @returns {Map}
       */
export const setValue = (destination, key, value) => destination.set(key, value);
/**
       * @param {Map} source source map object to access. Currently expects a Map, but could be refactored to use any type
       * @param {string} key
       * @returns {Map}
       */
export const getValue = (source, key) => source.get(key);
