// @ciscospark/internal-plugin-ediscovery-report-client/src/ediscovery-report-client.js
/* eslint-disable prefer-template */
/*!
 * Copyright (c) 2015-2017 Cisco Systems, Inc. See LICENSE file.
 */

import {waitForValue, SparkPlugin} from '@ciscospark/spark-core';

import reportRequest from './report-request';

const EDiscoveryReportClient = SparkPlugin.extend({
  namespace: 'EDiscoveryReportClient',

  @waitForValue('@')
  createReport(userIds, keywords, spaceNames, spaceIds, range) {
    const body = reportRequest;
    body.userIds = userIds;
    body.keywords = keywords;
    body.spaceNames = spaceNames;
    body.spaceIds = spaceIds;
    body.range = range;

    return this.request({
      method: 'POST',
      service: 'ediscovery',
      resource: 'reports',
      body
    }).then((res) => {
      console.log(res);
      return res;
    }).catch((reason) => {
      console.log(reason);
      return reason;
    });
  },

  @waitForValue('@')
  getReports() {
    return this.request({
      method: 'GET',
      service: 'ediscovery',
      resource: 'reports'
    }).then((res) => {
      console.log(res);
      return res;
    }).catch((reason) => {
      console.log(reason);
      return reason;
    });
  }
});

export default EDiscoveryReportClient;
