// @webex/plugin-meetings/src/meeting/state.js

import StateMachine from 'javascript-state-machine';
import StateMachineHistory from 'javascript-state-machine/lib/history';

import {MEETING_STATE_MACHINE} from '../constants';

const MeetingStateMachine = {
  create() {
    return new StateMachine({
      init: MEETING_STATE_MACHINE.STATES.IDLE,
      transitions: [
        {
          name: MEETING_STATE_MACHINE.TRANSITIONS.START,
          from: MEETING_STATE_MACHINE.STATES.IDLE,
          to: MEETING_STATE_MACHINE.STATES.ON_GOING
        },
        {
          name: MEETING_STATE_MACHINE.TRANSITIONS.START,
          from: MEETING_STATE_MACHINE.STATES.IDLE,
          to: MEETING_STATE_MACHINE.STATES.INCOMING
        },
        {
          name: MEETING_STATE_MACHINE.TRANSITIONS.JOIN,
          from: MEETING_STATE_MACHINE.STATES.IDLE,
          to: MEETING_STATE_MACHINE.STATES.DIALING
        },
        {
          name: MEETING_STATE_MACHINE.TRANSITIONS.JOIN,
          from: MEETING_STATE_MACHINE.STATES.ON_GOING,
          to: MEETING_STATE_MACHINE.STATES.DIALING
        },
        {
          name: MEETING_STATE_MACHINE.TRANSITIONS.JOIN,
          from: MEETING_STATE_MACHINE.STATES.INCOMING,
          to: MEETING_STATE_MACHINE.STATES.DIALING
        },
        {
          name: MEETING_STATE_MACHINE.TRANSITIONS.LOCAL,
          from: MEETING_STATE_MACHINE.STATES.DIALING,
          to: MEETING_STATE_MACHINE.STATES.CONNECTING_MEDIA_LOCAL
        },
        {
          name: MEETING_STATE_MACHINE.TRANSITIONS.CONNECT,
          from: MEETING_STATE_MACHINE.STATES.CONNECTING_MEDIA_LOCAL,
          to: MEETING_STATE_MACHINE.STATES.CONNECTING_MEDIA_REMOTE
        },
        {
          name: MEETING_STATE_MACHINE.TRANSITIONS.ESTABLISH,
          from: MEETING_STATE_MACHINE.STATES.CONNECTING_MEDIA_REMOTE,
          to: MEETING_STATE_MACHINE.STATES.ESTABLISHED_MEDIA
        },
        {
          name: MEETING_STATE_MACHINE.TRANSITIONS.LEAVE,
          from: MEETING_STATE_MACHINE.STATES.ESTABLISHED_MEDIA,
          to: MEETING_STATE_MACHINE.STATES.TERMINATING
        },
        {
          name: MEETING_STATE_MACHINE.TRANSITIONS.END,
          from: MEETING_STATE_MACHINE.STATES.TERMINATING,
          to: MEETING_STATE_MACHINE.STATES.ENDED
        },
        {
          name: MEETING_STATE_MACHINE.TRANSITIONS.CLEAN,
          from: MEETING_STATE_MACHINE.STATES.ENDED,
          to: MEETING_STATE_MACHINE.STATES.IDLE
        },
        {
          name: MEETING_STATE_MACHINE.TRANSITIONS.CLEAN,
          from: MEETING_STATE_MACHINE.STATES.ERROR,
          to: MEETING_STATE_MACHINE.STATES.IDLE
        },
        {
          name: MEETING_STATE_MACHINE.TRANSITIONS.ERROR,
          from: '*',
          to: MEETING_STATE_MACHINE.STATES.ERROR
        }
      ],
      methods: {
        // TODO more dynamic logger functions
        onInit: (transition) => {
          console.log(
            `Meeting:index#onInit->${transition.event} fired! State changed from '${transition.from}' to '${
              transition.to
            }' with transition '${transition.transition}' on Meeting Object: '${this.id}'.`
          );
        },
        onError: (transition) => {
          console.log(
            `Meeting:index#onError->${transition.event} fired! State changed from '${transition.from}' to '${
              transition.to
            }' with transition '${transition.transition}' on Meeting Object: '${this.id}'.`
          );
        },
        onClean: (transition) => {
          console.log(
            `Meeting:index#onClean->${transition.event} fired! State changed from '${transition.from}' to '${
              transition.to
            }' with transition '${transition.transition}' on Meeting Object: '${this.id}'.`
          );
        },
        onJoin: (transition) => {
          console.log(
            `Meeting:index#onJoin->${transition.event} fired! State changed from '${transition.from}' to '${
              transition.to
            }' with transition '${transition.transition}' on Meeting Object: '${this.id}'.`
          );
        },
        onLocal: (transition) => {
          console.log(
            `Meeting:index#onLocal->${transition.event} fired! State changed from '${transition.from}' to '${
              transition.to
            }' with transition '${transition.transition}' on Meeting Object: '${this.id}'.`
          );
        },
        onConnect: (transition) => {
          console.log(
            `Meeting:index#onConnect->${transition.event} fired! State changed from '${transition.from}' to '${
              transition.to
            }' with transition '${transition.transition}' on Meeting Object: '${this.id}'.`
          );
        },
        onEstablish: (transition) => {
          console.log(
            `Meeting:index#onEstablish->${transition.event} fired! State changed from '${transition.from}' to '${
              transition.to
            }' with transition '${transition.transition}' on Meeting Object: '${this.id}'.`
          );
        }
        // TODO add all on<Transition> methods
        // TODO potentially use the state transition methods to actually initiate the steps, rather than just track state
        /* TODO
      onInvalidTransition: (transition, from, to) => { // TODO
          if (transition && from && to) {
          throw new Error('transition not allowed from that state');
          }
      },
      onPendingTransition: (transition, from, to) => { // TODO
          if (transition && from && to) {
          throw new Error('transition already in progress');
          }
      } */
      },
      plugins: [
        new StateMachineHistory({max: 25})
      ]
    });
  }  
};

export default MeetingStateMachine;
