// @ciscospark/internal-plugin-reaction/src/reaction.js
/*!
 * Copyright (c) 2015-2017 Cisco Systems, Inc. See LICENSE file.
 */

import {SparkPlugin} from '@ciscospark/spark-core';
import uuid from 'uuid';

/**
 * @typedef {Object} ReactionObject
 * @property {object} participant - an actor with an id and entryUUID
 * @property {string} activityID - the id of the activity being reacted upon
 * @property {string} convoID - the id of the target conversation
 * @property {string} targetURL - url of the target conversation
 * @property {string} reactionID - (optional) the id of the reaction to be changed
 */

const Reaction = SparkPlugin.extend({
  namespace: 'Reaction',

  /**
  * helper function to create the request object
  * @param {Object} options
  * @returns {Object} the reaction request object
  */
  createReactionRequestBody(options) {
    return {
      clientTempId: uuid.v4(),
      objectType: 'activity',
      verb: options.verb,
      actor: {
        objectType: 'person',
        id: options.participant.id,
        entryUUID: options.participant.entryUUID
      },
      object: {
        objectType: 'messageProperty',
        tags: ['REACTIONS'],
        id: options.activityID,
        reactionType: options.reactionID
      },
      target: {
        id: options.convoID,
        url: options.targetURL,
        objectType: 'conversation'
      }
    };
  },

  /**
  * helper function to validate options parameter
  * @param {Object} options
  * @returns {string} the rejection error message or null if options is good
  */
  validateReaction(options) {
    if (!options) {
      return '`options` is required';
    }
    if (!options.participant) {
      return '`options.participant` is required';
    }
    if (!options.activityID) {
      return '`options.activityID` is required';
    }
    if (!options.convoID) {
      return '`options.convoID` is required';
    }
    if (!options.targetURL) {
      return '`options.targetURL` is required';
    }
    return null;
  },

  /**
  * Add a reaction
  * @param {Object} reaction
  * @returns {Promise<Object>} Resolves with the reaction just added
  */
  add(reaction) {
    const reactionNotValidated = this.validateReaction(reaction);
    if (reactionNotValidated) {
      return Promise.reject(new Error(reactionNotValidated));
    }

    const reactionParams = reaction;
    reactionParams.verb = 'add';
    reactionParams.reactionID = '';

    return this.request({
      method: 'POST',
      service: 'conversation',
      resource: '/activities',
      body: this.createReactionRequestBody(reactionParams)
    })
      .then((res) => res.body);
  },

  /**
  * Add a reaction
  * @param {Object} reaction
  * @returns {Promise<Object>} Resolves with the reaction just added
  */
  delete(reaction) {
    const reactionNotValidated = this.validateReaction(reaction);
    if (reactionNotValidated) {
      return Promise.reject(new Error(reactionNotValidated));
    }
    if (!reaction.reactionID) {
      return Promise.reject(new Error('`reaction.reactionID` is required'));
    }

    const reactionParams = reaction;
    reactionParams.verb = 'delete';

    return this.request({
      method: 'POST',
      service: 'conversation',
      resource: '/activities',
      body: this.createReactionRequestBody(reactionParams)
    })
      .then((res) => res.body);
  }

});

export default Reaction;
