// ciscospark/src/plugins/webhooks.js
/*!
 * Copyright (c) 2015-2017 Cisco Systems, Inc. See LICENSE file.
 */

import {SparkPlugin, Page} from '@ciscospark/spark-core';

/**
 * @typedef {Object} WebhookObject
 * @property {string} id - The unique ID for the webhook.
 * @property {string} resource - The resource type for the webhook.
 * @property {string} event - The event type for the webhook.
 * @property {string} filter - The filter that defines the webhook scope.
 * @property {string} targetUrl - The URL that receives POST requests for each event.
 * @property {string} name - A user-friendly name for this webhook.
 */

/**
 * A webhook notifies an application when an event for which the application is
 * registered has occurred.
 * @class
 */
const Webhooks = SparkPlugin.extend({
  /**
   * Posts a webhook.
   * @instance
   * @memberof Webhooks
   * @param {WebhookObject} webhook
   * @returns {Promise<Webhook>}
   * @example
   * ciscospark.rooms.create({title: 'Create Webhook Example'})
   *   .then(function(room) {
   *     return ciscospark.webhooks.create({
   *       resource: 'messages',
   *       event: 'created',
   *       filter: 'roomId=' + room.id,
   *       targetUrl: 'https://example.com/webhook',
   *       name: 'Test Webhook'
   *     });
   *   })
   *   .then(function(webhook) {
   *     return 'success';
   *   });
   *   // => success
   */
  create(webhook) {
    return this.request({
      method: `POST`,
      service: `hydra`,
      resource: `webhooks`,
      body: webhook
    })
      .then((res) => res.body);
  },

  /**
   * Shows details for a webhook.
   * @instance
   * @memberof Webhooks
   * @param {Webhook|string} webhook
   * @returns {Promise<Array<Webhook>>}
   * @example
   * ciscospark.rooms.create({title: 'Get Webhook Example'})
   *   .then(function(room) {
   *     return ciscospark.webhooks.create({
   *       resource: 'messages',
   *       event: 'created',
   *       filter: 'roomId=' + room.id,
   *       targetUrl: 'https://example.com/webhook',
   *       name: 'Test Webhook'
   *     });
   *   })
   *   .then(function(webhook) {
   *     return ciscospark.webhooks.get(webhook.id);
   *   })
   *   .then(function(webhook2) {
   *     return 'success';
   *   });
   *   // => success
   */
  get(webhook) {
    const id = webhook.id || webhook;

    return this.request({
      service: `hydra`,
      resource: `webhooks/${id}`
    })
      .then((res) => res.body.items || res.body);
  },

  /**
   * Lists all webhooks.
   * @instance
   * @memberof Webhooks
   * @param {Object} options
   * @param {integer} options.max Limit the maximum number of webhooks in the response.
   * @returns {Promise<Webhook>}
   * @example
   * var room, webhook;
   * ciscospark.rooms.create({title: 'List Webhooks Example'})
   *   .then(function(r) {
   *     room = r;
   *     return ciscospark.webhooks.create({
   *       resource: 'messages',
   *       event: 'created',
   *       filter: 'roomId=' + room.id,
   *       targetUrl: 'https://example.com/webhook',
   *       name: 'Test Webhook'
   *     });
   *   })
   *   .then(function(webhook) {
   *     return ciscospark.webhooks.list();
   *   })
   *   .then(function(webhooks) {
   *     return 'success';
   *   });
   *   // => success
   */
  list(options) {
    return this.request({
      service: `hydra`,
      resource: `webhooks/`,
      qs: options
    })
      .then((res) => new Page(res, this.spark));
  },

  /**
   * Delete a webhook.
   * @instance
   * @memberof Webhooks
   * @param {Webhook|string} webhook
   * @returns {Promise}
   * @example
   * var room, webhook;
   * ciscospark.rooms.create({title: 'Remove Webhook Example'})
   *   .then(function(r) {
   *     room = r;
   *     return ciscospark.webhooks.create({
   *       resource: 'messages',
   *       event: 'created',
   *       filter: 'roomId=' + room.id,
   *       targetUrl: 'https://example.com/webhook',
   *       name: 'Test Webhook'
   *     });
   *   })
   *   .then(function(webhook) {
   *     return ciscospark.webhooks.remove(webhook);
   *   })
   *   .then(function() {
   *     return ciscospark.webhooks.list();
   *   })
   *   .then(function(webhooks) {
   *     return 'success';
   *   });
   *   // => success
   */
  remove(webhook) {
    const id = webhook.id || webhook;
    return this.request({
      method: `DELETE`,
      service: `hydra`,
      resource: `webhooks/${id}`
    })
      .then((res) => {
        // Firefox has some issues with 204s and/or DELETE. This should move to
        // http-core
        if (res.statusCode === 204) {
          return undefined;
        }
        return res.body;
      });
  },

  /**
   * Update a webhook.
   * @instance
   * @memberof Webhooks
   * @param {Webhook} webhook
   * @returns {Promise<Webhook>}
   * @example
   * var webhook;
   * ciscospark.rooms.create({title: 'Webhook Example'})
   *   .then(function(room) {
   *     return ciscospark.webhooks.create({
   *       resource: 'messages',
   *       event: 'created',
   *       filter: 'roomId=' + room.id,
   *       targetUrl: 'https://example.com/webhook',
   *       name: 'Test Webhook'
   *     });
   *   })
   *   .then(function(w) {
   *     webhook = w;
   *     webhook.targetUrl = 'https://example.com/webhook/newtarget';
   *     return ciscospark.webhooks.update(webhook);
   *   })
   *   .then(function() {
   *     return ciscospark.webhooks.get(webhook);
   *   })
   *   .then(function(webhook) {
   *     return 'success';
   *   });
   *   // => success
   */
  update(webhook) {
    const id = webhook.id;
    return this.request({
      method: `PUT`,
      service: `hydra`,
      resource: `webhooks/${id}`,
      body: webhook
    })
      .then((res) => res.body);
  }
});

export default Webhooks;
