// @webex/plugin-meetings/src/roap/constants.js
/*!
 * Copyright (c) 2015-2018 Cisco Systems, Inc. See LICENSE file.
 */

export const roapTypes = {
  OFFER: 'OFFER',
  ANSWER: 'ANSWER',
  OK: 'OK',
  ERROR: 'ERROR',
  SHUTDOWN: 'SHUTDOWN',
  OFFER_REQUEST: 'OFFER_REQUEST'
};

export const roapState = {
  INIT: 'INIT',
  WAIT_RX_OFFER: 'WAIT_RX_OFFER',
  WAIT_RX_ANSWER: 'WAIT_RX_ANSWER',
  WAIT_RX_OK: 'WAIT_RX_OK',
  WAIT_TX_OFFER: 'WAIT_TX_OFFER',
  WAIT_TX_ANSWER: 'WAIT_TX_ANSWER',
  WAIT_TX_OK: 'WAIT_TX_OK',
  IDLE_LOCAL_OFFER: 'IDLE_LOCAL_OFFER',
  IDLE_REMOTE_OFFER: 'IDLE_REMOTE_OFFER',
  GLARE: 'GLARE',
  ERROR: 'ERROR'
};

export const roapSignal = {
  RX_OFFER: 'RX_OFFER',
  TX_OFFER: 'TX_OFFER',
  RX_ANSWER: 'RX_ANSWER',
  TX_ANSWER: 'TX_ANSWER',
  RX_OK: 'RX_OK',
  TX_OK: 'TX_OK',
  GLARE_RESOLVED: 'GLARE_RESOLVED'
};

export const roapMessages = {
  RECEIVE_ROAP_MSG: 'RECEIVE_ROAP_MSG',
  SEND_ROAP_MSG: 'SEND_ROAP_MSG',
  SEND_ROAP_MSG_SUCCESS: 'SEND_ROAP_MSG_SUCCESS',
  RESET_ROAP_STATE: 'RESET_ROAP_STATE',
  RECEIVE_CALL_LEAVE: 'RECEIVE_CALL_LEAVE'
};
