// @ciscospark/plugin-lyra/src/lyra.js
/**!
 *
 * Copyright (c) 2015-2017 Cisco Systems, Inc. See LICENSE file.
 * @private
 */

import {SparkPlugin} from '@ciscospark/spark-core';
import Spaces from './spaces';

const Lyra = SparkPlugin.extend({
  namespace: `Lyra`,
  children: {
    spaces: Spaces
  },

  /**
   * Pings lyra
   * @returns {Promise<Object>} ping response body
   */
  ping() {
    return this.spark.request({
      method: `GET`,
      api: `lyra`,
      resource: `/ping`
    })
      .then((res) => res.body);
  },

  /**
   * Gets lyra service health
   * @returns {Promise<Object>} healthStatus response body
   */
  getServiceHealth() {
    return this.spark.request({
      method: `GET`,
      api: `lyra`,
      resource: `/service_health`
    })
      .then((res) => res.body);
  },

  /**
   * Gets lyra build info
   * @returns {Promise<Object>} buildInfo response body
   */
  getBuildInfo() {
    return this.spark.request({
      method: `GET`,
      api: `lyra`,
      resource: `/build_info`
    })
      .then((res) => res.body);
  }

});


export default Lyra;
