// @webex/plugin-meetings/src/participants/util.js
import {PUT, CONTROLS} from '../constants';

const ParticipantUtil = {};

ParticipantUtil.generateAddParticipantOptions = (invitee, locusUrl, alertIfActive) => ({
  invitee,
  locusUrl,
  alertIfActive
});

ParticipantUtil.generateAdmitParticipantOptions = (participantIds, locusUrl) => ({
  locusUrl,
  participantIds
});

ParticipantUtil.getAddParticipantBody = (options) => ({
  invitees: [
    {
      address: options.invitee.emailAddress || options.invitee.email
    }
  ],
  alertIfActive: options.alertIfActive
});

ParticipantUtil.getAdmitParticipantRequestBody = (options) => ({
  admit: {participantIds: options.participantIds}
});

ParticipantUtil.getAdmitParticipantRequestParams = (format) => {
  const body = ParticipantUtil.getAdmitParticipantRequestBody(format);
  return {
    method: PUT,
    uri: `${format.locusUrl}/${CONTROLS}`,
    body
  };
};

ParticipantUtil.getAddParticipantRequestParams = (format) => {
  const body = ParticipantUtil.getAddParticipantBody(format);
  const requestParams = {
    method: PUT,
    uri: format.locusUrl,
    body
  };
  return requestParams;
};

ParticipantUtil.extractAddParticipant = (body) => ({id: 'test', body});

export default ParticipantUtil;
