// @webex/plugin-meetings/src/meeting-info/request.js
import {StatelessSparkPlugin} from '@ciscospark/spark-core';
import btoa from 'btoa';

import {
  SIP_URI,
  PERSONAL_ROOM,
  MEETING_ID,
  CONVERSATION_URL,
  LOCUS_ID,
  MEETING_LINK,
  WEBEX_DOT_COM,
  LOCUS,
  GET,
  PUT,
  USE_URI_LOOKUP_FALSE,
  TYPE,
  LOCI,
  MEETINGINFO,
  MEETINGS,
  ALTERNATE_REDIRECT_TRUE
} from '../constants';

/**
 * Helper function to check if a string matches a known meeting link pattern
 * @param {String} value  string to parse and see if it matches a meeting link
 * @returns {String} returns a sipURL
 * @returns {null} returns null if it's not a valid meeting link
 */
function checkforMeetingLink(value) {
  // eslint-disable-next-line no-warning-comments
  const regex = /\/\/(.*).webex.com\/meet\/(.*)/; // TODO lets remove regex from this equation and user URI matchers and such
  const m = regex.exec(value);
  let sipUrl = null;

  if (m !== null && m.length === 3) {
    // The result can be accessed through the `m`-variable.
    sipUrl = `${m[2]}@${m[1]}.${WEBEX_DOT_COM}`;
  }

  return sipUrl;
}

/**
 * Helper function to build up a correct locus url depending on the value passed
 * @param {String} type One of [SIP_URI, PERSONAL_ROOM, MEETING_ID, CONVERSATION_URL, LOCUS_ID, MEETING_LINK]
 * @param {Object} value ?? value.value
 * @returns {Object} returns an object with {resource, method}
 */
function getResourceUrl(type, value) {
  let resource = `/${LOCI}/${MEETINGINFO}`;
  let method = GET;
  switch (type) {
    case SIP_URI:
    case PERSONAL_ROOM:
    case MEETING_ID:
      resource = `/${LOCI}/${MEETINGINFO}/${value.value}?${TYPE}=${type}&${USE_URI_LOOKUP_FALSE}`;
      break;
    case CONVERSATION_URL:
      method = PUT;
      break;
    case LOCUS_ID:
      // if locus id just send PUT request
      resource = `/${LOCI}/${value.value}/${MEETINGINFO}`;
      method = PUT;
      break;
    case MEETING_LINK:
      resource = `$/${LOCI}/${MEETINGINFO}/${btoa(value.value)}?${TYPE}=${MEETING_LINK}&${USE_URI_LOOKUP_FALSE}`;
      break;
    default:
  }
  return {
    resource,
    method
  };
}
/**
 * @class MeetingInfoRequest
 */
export default class MeetingInfoRequest extends StatelessSparkPlugin {
  namespace = MEETINGS;

  /**
   *
   * @param {String} destination One of many types of destinations. Currently only sip uri works
   * @returns {Promise} returns a promise that resolves/rejects the result of the request
   */
  fetchMeetingInfo(destination) {
    let value = destination;
    let type = SIP_URI;

    // Making sure all the webEx meeting are converted into SipUrl
    const sipUrl = checkforMeetingLink(value);

    if (sipUrl) {
      type = SIP_URI;
      value = sipUrl;
    }

    const resourceOptions = getResourceUrl(type, value);

    const requestParams = {
      method: resourceOptions.method,
      api: LOCUS,
      resource: resourceOptions.resource
    };

    if (resourceOptions.method === GET) {
      // for handling URL redirections
      requestParams.resource = requestParams.resource.concat(`&${ALTERNATE_REDIRECT_TRUE}`);
    }
    else if (type !== LOCUS_ID) {
      // locus id check is a PUT not sure why
      requestParams.body = {
        value,
        lookupType: type,
        alternateRedirect: true
      };
    }

    return this.request(requestParams);
  }
}
