// @ciscospark/internal-plugin-ediscovery/src/report-generator.js
import SparkPlugin from '@ciscospark/spark-core';
import JSZip from 'jszip';
import saveAs from 'file-saver';
import MimeBuilder from 'emailjs-mime-builder';

/**
 * @class ReportGenerator is used to compile, zip, and save an eDiscovery report to disk.
 *
 * Warning: This file is browser specific. It cannot function in a NodeJS server environment.
 */
const ReportGenerator = SparkPlugin.extend({
  namespace: 'Ediscovery',

  props: {
    zip: 'object',
    folder: 'object'
  },

  /**
   * Map an activity to an MIME rfc822 formated string that can be saved as an eml file.
   * @param {Object} activity An activity object
   * @returns {String} A string that represents the data stored in an eml file.
   */
  map(activity) {
    if (!activity) {
      throw new Error('Invalid parameter(s)');
    }

    return new MimeBuilder('multipart/mixed')
      .setHeader('From', activity.actorId)
      .setHeader('To', 'Foo Bar <foo@bar.com>')
      .setHeader('Subject', activity.activityId)
      .setHeader('Date', activity.timestamp)
      .setContent(activity.content)
      .build();
  },

  /**
   * Add activities to the report.
   * @param {String} reportId An id representing the report
   * @param {Object[]} activities Activities that are to form part of the report
   * @returns {Void}
   */
  add(reportId, activities) {
    if (!reportId || !activities) {
      throw new Error('Invalid parameter(s)');
    }

    if (!this.zip) {
      this.zip = new JSZip();
      this.folder = this.zip.folder(`${reportId}/conversation`);
    }

    this.logger.info('Creating Folder Structure');

    for (let i = 0; i < activities.length; i += 1) {
      const activity = activities[i];
      const data = this.map(activity);
      this.folder.file(`${activity.activityId}.eml`, data);
    }
  },

  /**
   * Once all the activities have been added create a report and save it to disk.
   * @param {String} reportId An id representing the report
   * @returns {Promise} A promise on the generated report
   */
  saveAs(reportId) {
    if (!reportId) {
      throw new Error('Invalid parameter');
    }

    if (!this.zip) {
      throw new Error('No activites added to report');
    }

    this.logger.info('Creating Zip File');
    return this.zip.generateAsync({type: 'blob'})
      .then((content) => {
        // Download to the user's browser
        this.logger.info('Downloading...');
        saveAs(content, `${reportId}.zip`);
      });
  }
});

export default ReportGenerator;
