// @ciscospark/internal-plugin-encryption/src/config.js
/*!
 *
 * Copyright (c) 2015-2017 Cisco Systems, Inc. See LICENSE file.
 */

/** @typedef {Object} config */

export default {
  /**
   * @memberof config
   */
  encryption: {
    /**
     * @memberof config.encryption
     */
    joseOptions: {
      /**
       * @memberof config.encryption.joseOptions
       */
      compact: true,
      /**
       * @memberof config.encryption.joseOptions
       */
      contentAlg: `A256GCM`,
      /**
       * @memberof config.encryption.joseOptions
       */
      protect: `*`
    },

    /**
     * @memberof config.encryption
     */
    kmsInitialTimeout: 6000,

    /**
     * @memberof config.encryption
     */
    kmsMaxTimeout: 32000,

    /**
     * Debounce wait before sending a kms request
     * @memberof config.encryption
     * @type {number}
     */
    batcherWait: 50,

    /**
     * Maximum queue size before sending a kms request
     * @memberof config.encryption
     * @type {number}
     */
    batcherMaxCalls: 50,

    /**
     * Debounce max wait before sending a kms request
     * @memberof config.encryption
     * @type {number}
     */
    batcherMaxWait: 150
  }
};
