// @ciscospark/internal-plugin-encryption/src/kms-errors.js
/*!
 * Copyright (c) 2015-2017 Cisco Systems, Inc. See LICENSE file.
 */

import {Exception} from '@ciscospark/common';
import {SparkHttpError} from '@ciscospark/spark-core';

/**
 * Base class for KMS errors
 * @param {HttpResponse|Object} response
 */
export class KmsError extends Exception {
  static defaultMessage = 'An unknown error occurred while communicating with the kms. This implies we received an error response without a body.';

  /**
   * @param {HttpResponse} body
   * @private
   * @returns {string}
   */
  parse(body) {
    const actualBody = body.body || body;

    Object.defineProperties(this, {
      body: {
        enumerable: false,
        value: actualBody
      },
      reason: {
        enumerable: false,
        value: actualBody.reason
      },
      requestId: {
        enumerable: false,
        value: actualBody.requestId
      },
      status: {
        enumerable: false,
        value: actualBody.status
      }
    });

    let message = typeof actualBody === 'string' ? actualBody : actualBody.reason;
    if (!message) {
      message = this.constructor.defaultMessage;
    }
    if (actualBody.status) {
      message += `\nKMS_RESPONSE_STATUS: ${actualBody.status}`;
    }
    if (actualBody.requestId) {
      message += `\nKMS_REQUEST_ID: ${actualBody.requestId}`;
    }

    return message;
  }
}

/**
 * Thrown when an expected KMSResponse is not received in a timely manner
 * @param {Object} options
 * @param {KmsRequest} options.request
 * @param {number} options.timeout
 */
export class KmsTimeoutError extends KmsError {
  /**
   * @param {KmsRequest} options.request
   * @param {number} options.timeout
   * @private
   * @returns {string}
   */
  parse({request = {}, timeout} = {}) {
    let message = `The KMS did not respond within ${timeout ? `${timeout} milliseconds` : 'a timely fashion'}`;
    if (request) {
      if (request.method && request.uri) {
        message += `\nKMS_REQUEST: ${request.method} ${request.uri}`;
      }

      if (request.requestId) {
        message += `\nKMS_REQUEST_ID: ${request.requestId}`;
      }
    }

    return message;
  }
}

/**
 * Emitted when a REST request includes an encrypter error
 * @param {SparkHttpError} response
 */
export class DryError extends SparkHttpError {
  static defaultMessage = 'An unknown error was received from a service that proxies to the KMS';

  /**
   * @param {SparkHttpError} reason
   * @private
   * @returns {string}
   */
  parse(reason) {
    Reflect.apply(SparkHttpError.prototype.parse, this, [reason._res]);
    const body = reason._res.body.message;

    let message = body.reason || body;
    if (!message) {
      message = this.constructor.defaultMessage;
    }
    if (this.options.url) {
      message += `\n${this.options.method} ${this.options.url}`;
    }
    else if (this.options.uri) {
      message += `\n${this.options.method} ${this.options.uri}`;
    }
    else {
      message += `\n${this.options.method} ${this.options.service.toUpperCase()}/${this.options.resource}`;
    }
    message += `\nWEBEX_TRACKING_ID: ${this.options.headers.trackingid}`;

    if (body.status) {
      message += `\nKMS_RESPONSE_STATUS: ${body.status}`;
    }
    if (body.requestId) {
      message += `\nKMS_REQUEST_ID: ${body.requestId}`;
    }

    Object.defineProperties(this, {
      reason: {
        enumerable: false,
        value: body.reason
      },
      requestId: {
        enumerable: false,
        value: body.requestId
      },
      status: {
        enumerable: false,
        value: body.status
      }
    });

    return message;
  }
}
