// @ciscospark/plugin-phone/browsers.js
// eslint-disable-next-line strict
'use strict';

/* eslint camelcase: [0] */
const path = require(`path`);
module.exports = function createBrowsers() {
  const browsers = {
    local: {
      firefox_h264: {
        base: `Firefox`
        // profile: path.resolve(process.cwd(), `.tmp/selenium/mac`)
      }
      // ChromeH264: {
      //   base: `Chrome`,
      //   flags: [
      //     `--use-fake-device-for-media-stream`,
      //     `--use-fake-ui-for-media-stream`
      //   ]
      // }
    },
    sauce: {
      // sl_chrome_latest_osx12: {
      //   base: `SauceLabs`,
      //   platform: `OS X 10.12`,
      //   browserName: `chrome`,
      //   version: `latest`,
      //   chromeOptions: {
      //     args: [
      //       `--use-fake-device-for-media-stream`,
      //       `--use-fake-ui-for-media-stream`
      //     ]
      //   }
      // },
      // sl_firefox_latest_linux: {
      //   base: `SauceLabs`,
      //   platform: `Linux`,
      //   browserName: `firefox`,
      //   version: `latest`
      // },
      sl_firefox_latest_osx12: {
        base: `SauceLabs`,
        platform: `OS X 10.12`,
        browserName: `firefox`,
        version: `latest`,
        prerun: `sauce-storage:openh264.sh`
      }
      // sl_firefox_latest_win7: {
      //   base: `SauceLabs`,
      //   platform: `Windows 7`,
      //   browserName: `firefox`,
      //   version: `latest`
      // }
    }
  };

  return browsers;
};
