// @webex/plugin-meetings/src/locus-Info/utils/self.js
/* globals spark */
const selfUtils = {};

selfUtils.parse = (self) => ({
  muted: self.controls.audio.muted,
  guest: self.guest,
  moderator: self.moderator,
  mediaStatus: selfUtils.getStatus(self.status),
  state: self.state,
  // TODO: give a proper name . With same device as login or different login`
  // Some times we might have joined with both mobile and web
  joinedWith: selfUtils.checkJoinedStatus(self) ? 'SELF' : 'OTHER',
  creator: self.isCreator // check if its used
});

/**
 * returns true of false if this device has joined the meeting
 * @param {boolen} self  true
 * @returns {boolen} return true if you have joined the meeting
 */
selfUtils.checkJoinedStatus = (self) => {
  // Device you logged in with
  const selfDeviceUrl = spark.internal.device.url;
  return self.devices.some((device) => selfDeviceUrl === device.Url);
};

selfUtils.getStatus = (status) => ({
  audio: status.audioStatus,
  video: status.videoStatus,
  slides: status.videoSlidesStatus
});


export default selfUtils;
