// @webex/plugin-meetings/src/config.js
export default {
  // TODO: this needs to be defaulted
  // to JS_SDK and clients set it to WEB or others
  // see https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-73167
  meetings: {
    deviceType: 'WEB',
    mediaSettings: {
      sendAudio: true,
      sendVideo: true,
      receiveAudio: true,
      receiveVideo: true,
      pstn: false,
      sendShare: false,
      receiveShare: false
    },
    reconnection: {
      enabled: true,
      detection: true,
      retry: {
        times: 2,
        backOff: {
          start: 1000,
          rate: 2
        }
      }
    },
    stats: {
      // 1 second intervals to collect stats data
      interval: 1000,
      // hold the last 2 minute of a calls data
      historyMax: 120,
      // send a metrics event automatically
      autoSendMediaDiagnostics: false,
      // dispatch data every 10 seconds
      // so that metrics consumer doesn't get flooded
      mediaDiagnosticsInterval: 10
    },
    logging: {
      enable: true,
      verboseEvents: true
    },
    resolution: {
      maxWidth: 1280,
      maxHeight: 720,
      idealWidth: 1280,
      idealHeight: 720
    },
    screenResolution: {
      maxWidth: 1920,
      maxHeight: 1080,
      idealWidth: 1920,
      idealHeight: 1080
    },
    bandwidth: {
      // please note, these are the maximum bandwidth values
      // the server supports, minimums have to be tested
      audio: 64000,
      video: 4000000
    },
    screenFrameRate: 10,
    videoShareFrameRate: 30,
    aspectRatio: 1.7695852534562213
  }
};
