// @webex/plugin-meetings/src/config.js
export default {
  meetings: {
    mediaSettings: {
      sendAudio: true,
      sendVideo: true,
      receiveAudio: true,
      receiveVideo: true,
      pstn: false,
      sendShare: false,
      receiveShare: false
    },
    reconnection: {
      enabled: true,
      detection: true,
      retry: {
        times: 2,
        backOff: {
          start: 1000,
          rate: 2
        }
      }
    },
    stats: {
      senders: {
        audio: {
          transceiver: 'audioTransceiver',
          child: 'sender',
          parent: 'mediaProperties',
          peerConnection: 'peerConnection',
          name: 'audioSender'
        },
        video: {
          transceiver: 'videoTransceiver',
          child: 'sender',
          parent: 'mediaProperties',
          peerConnection: 'peerConnection',
          name: 'videoSender'
        },
        share: {
          transceiver: 'shareTransceiver',
          child: 'sender',
          parent: 'mediaProperties',
          peerConnection: 'peerConnection',
          name: 'shareSender'
        }
      },
      receivers: {
        audio: {
          transceiver: 'audioTransceiver',
          child: 'receiver',
          parent: 'mediaProperties',
          peerConnection: 'peerConnection',
          name: 'audioReceiver'
        },
        video: {
          transceiver: 'videoTransceiver',
          child: 'receiver',
          parent: 'mediaProperties',
          peerConnection: 'peerConnection',
          name: 'videoReceiver'
        },
        share: {
          transceiver: 'shareTransceiver',
          child: 'receiver',
          parent: 'mediaProperties',
          peerConnection: 'peerConnection',
          name: 'shareReceiver'
        }
      }
    }
  }
};
