// @webex/plugin-meetings/src/member.js
/*!
 * Copyright (c) 2015-2018 Cisco Systems, Inc. See LICENSE file.
 */

/**
 * Events to emit
 * ==============
 *
 * - change (string): audioMuted | videoMuted | left | joined | madeHost | shared
 * // NOTE: /calls API had: waiting | lobby | notified | declined | joined | left
 */
export default class Member {
  /**
   * Create a Member.
   * @param {string} id
   */
  constructor({
    id, isGuest, isHost, isAudioMuted, isVideoMuted, devices
  }) {
    this.id = id;
    this.isGuest = isGuest;
    this.isHost = isHost;
    this.isAudioMuted = isAudioMuted;
    this.isVideoMuted = isVideoMuted;
    this.devices = devices;
  }

  /**
   * Update member properties.
   * @returns {Promise}
   */
  async update() {
    return new Promise(async (resolve) => {
      const properties = {
        isGuest: false,
        isHost: false,
        isAudioMuted: false,
        isVideoMuted: false,
        devices: []
      };
      const member = await this.requestFromServer(properties);
      // TODO: Destructure with this.id
      resolve(member);
      return member;
    });
  }

  /**
   * Mute outgoing audio.
   * @returns {Promise}
   */
  async muteAudio() {
    return this.requestMuteAudio(true);
  }

  /**
   * Unmute outgoing audio.
   * @returns {Promise}
   */
  async unmuteAudio() {
    return this.requestMuteAudio(false);
  }

  /**
   * Request muting/unmuting outgoing audio.
   * @private
   * @param {boolean} mute
   * @returns {Promise}
   */
  async requestMuteAudio(mute) {
    return new Promise(async (resolve, reject) => {
      // Request to mute audio from the server.
      const success = await this.requestFromServer(mute);
      if (success) {
        // TODO: log success
        this.isAudioMuted = true;
        resolve(this.isAudioMuted);
        return this.isAudioMuted;
      }
      // TODO: log error
      return reject();
    });
  }

  /**
   * Mute outgoing video.
   * @returns {Promise}
   */
  async muteVideo() {
    return this.requestMuteVideo(true);
  }

  /**
   * Unmute outgoing video.
   * @returns {Promise}
   */
  async unmuteVideo() {
    return this.requestMuteVideo(false);
  }

  /**
   * Request muting/unmuting outgoing video.
   * @private
   * @param {boolean} mute
   * @returns {Promise}
   */
  async requestMuteVideo(mute) {
    return new Promise(async (resolve, reject) => {
      // Request to mute audio from the server.
      const success = await this.requestFromServer(mute);
      if (success) {
        // TODO: log success
        this.isVideoMuted = true;
        resolve(this.isVideoMuted);
        return this.isVideoMuted;
      }
      // TODO: log error
      return reject();
    });
  }

  /**
   * Resolve any given value.
   * TODO: Dummy method to illustrate server requests.
   * @private
   * @static
   * @param {*} value - can be anything
   * @returns {Promise}
   */
  async requestFromServer(value) {
    return value;
  }
}

/*

### _remove()_ - function
> For Webex, only Webex Hosts can remove. For Teams, anyone can remove.
- return Promise
    - resolve
    - reject
        - if something on server fails
        - if, on Webex, they are not the host

### _admit()_ - function
> Admit a member in the *Lobby*.
- return Promise

### _makeHost()_ - function
> For Webex, only Hosts can make another person a Host. For Teams,
> everyone has "host" capabilities.
- return Promise
    - resolve
    - reject
        - if something on server fails
        - if, on Webex, they are not the host

*/
