// @webex/plugin-meetings/src/personal-meeting-room/index.js
import {StatelessSparkPlugin} from '@ciscospark/spark-core';

import {MEETINGS, PERSONAL_ROOM} from '../constants';
import MeetingInfoRequest from '../meeting-info/request';

/**
 * @class PersonalMeetingRoom
 */
export default class PersonalMeetingRoom extends StatelessSparkPlugin {
  namespace = MEETINGS;

  /**
   *
   * @param {Object} attrs
   * @param {Object} options
   */
  constructor(attrs, options) {
    super({}, options);
    this.pmr = null;
    this.sipUri = null;
    this.link = null;
    this.userId = null;
    this.name = null;
    this.MeetingInfoRequest = new MeetingInfoRequest({}, options);
  }

  /**
   * @param {Object} body the response body from meeting info request
   * @returns {undefined}
   * @memberof PersonalMeetingRoom
   */
  set(body) {
    this.pmr = body;
    this.sipUri = body.sipMeetingUri;
    this.meetingLink = body.webExMeetingLink;
    this.userId = body.owner;
    this.name = body.meetingName;
    this.number = body.meetingNumber;
  }

  /**
   * TODO: implement TTL for syncing and caching so to not request again and again
   * @param {Object} options
   * @returns {Promise}
   * @memberof PersonalMeetingRoom
   */
  get() {
    const options = {
      type: PERSONAL_ROOM,
      destination: this.spark.internal.device.userId
    };
    return this.MeetingInfoRequest.fetchMeetingInfo(options).then((pmr) => {
      if (pmr.body && pmr.body.isPmr) {
        this.set(pmr.body);
      }
      else {
        return Promise.reject(new Error('The PMR requested is NOT a PMR'));
      }
      return pmr;
    });
  }
}
