// @webex/internal-plugin-ediscovery/test/unit/spec/activities.js
const Activities = [
  {
    activityId: '029e57c0-e694-11e8-98d1-b1349cfbb607',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/19288ee4-d235-4897-93f5-6cdd433ac420',
    objectDisplayName: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..I8dprtpOu6kk4NYO._gE9BUI6E4H4H2FzxStZWq81rJQ7RvEh1idkHtpFQMYC8TYtaN2D_Z053Q.s8mRZ1vME9C7Ne7JmQrgZQ',
    objectId: '',
    targetId: '223b3dbb-0980-34c0-8d4c-52820b6ee5c9',
    timestamp: '2018-11-12T16:00:00.509Z',
    verb: 'post'
  },
  {
    activityId: '14113360-e694-11e8-981c-f5e3fd7b0d6e',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    objectId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:00:30.102Z',
    verb: 'add'
  },
  {
    activityId: '14115a70-e694-11e8-981c-f5e3fd7b0d6e',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    objectId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:00:30.102Z',
    verb: 'assignModerator'
  },
  {
    activityId: '14b306c0-e696-11e8-98d1-b1349cfbb607',
    actorId: '7765121f-f04e-4bae-8f83-16ac80a3315b',
    actorDisplayName: 'Test User <testuser@email.com>',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/2a69ecbe-a630-4845-9221-818449d7bd7c',
    objectDisplayName: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..bmyKDuR84yTqQVdj.A8IPRqIc3g4.fsgwcN8pdUrHTr4JMf2u8Q',
    objectId: '',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:14:49.958Z',
    verb: 'post'
  },
  {
    activityId: '1b575030-e696-11e8-98d1-b1349cfbb607',
    actorId: '66e8b038-3a71-4857-9f42-65a730c93ab2',
    actorDisplayName: 'Test User <testuser@email.com>',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/2a69ecbe-a630-4845-9221-818449d7bd7c',
    objectDisplayName: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..Sg8Q5eItntFtv_6k.-oOiiOc.uAcNMWljkOsRmkvhAjvcHw',
    objectId: '',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:15:01.113Z',
    verb: 'post'
  },
  {
    activityId: '29691160-e694-11e8-981c-f5e3fd7b0d6e',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    objectId: '7765121f-f04e-4bae-8f83-16ac80a3315b',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:05.910Z',
    verb: 'add'
  },
  {
    activityId: '2b90fa70-e694-11e8-98d1-b1349cfbb607',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    objectId: 'a003e2e4-cdfb-4bfb-8a05-8109fe64ca99',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:09.527Z',
    verb: 'add'
  },
  {
    activityId: '2dfb1ca0-e694-11e8-981c-f5e3fd7b0d6e',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    objectId: '3ae352c3-f32c-4cf2-8385-8e03c4c801ed',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:13.578Z',
    verb: 'add'
  },
  {
    activityId: '37ba61b0-e694-11e8-981c-f5e3fd7b0d6e',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    objectId: '66e8b038-3a71-4857-9f42-65a730c93ab2',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:29.931Z',
    verb: 'add'
  },
  {
    activityId: '3c60b840-e694-11e8-981c-f5e3fd7b0d6e',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    objectId: 'cbf0323f-44e9-4b56-8514-a0a7ed3b333a',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:37.732Z',
    verb: 'add'
  },
  {
    activityId: '42f411c0-e694-11e8-981c-f5e3fd7b0d6e',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    objectId: '91facb42-c96f-441d-b45c-718ba5a71786',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:48.764Z',
    verb: 'add'
  },
  {
    activityId: '47a00da0-e694-11e8-98d1-b1349cfbb607',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    objectId: '93b4bd25-ed19-4c72-81a4-a9c8b6241708',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:56.602Z',
    verb: 'add'
  },
  {
    activityId: '4aac48e0-e768-11e8-9ad1-cf6f6822e300',
    actorId: '66e8b038-3a71-4857-9f42-65a730c93ab2',
    actorDisplayName: 'Test User <testuser@email.com>',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/2a69ecbe-a630-4845-9221-818449d7bd7c',
    objectDisplayName: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..YuNeE_v-jk0XIW7o.nhU.7Ku7utNFVaQ0FQ6qWIanSA',
    objectId: '',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-13T17:19:34.166Z',
    verb: 'post'
  },
  {
    activityId: '7a9ec280-e768-11e8-afdf-b54a3c174c2d',
    timestamp: '2018-11-13T17:20:55.464Z',
    verb: 'share',
    actorId: '66e8b038-3a71-4857-9f42-65a730c93ab2',
    actorDisplayName: 'Test User <testuser@email.com>',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/2a69ecbe-a630-4845-9221-818449d7bd7c',
    objectId: '',
    files: [
      {
        displayName: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..9Cu-nKDWs1zQ2bo1.quNXLgmoGdEKxH-6.sOSAeS7r7heP0xrrJmCphw',
        url: 'https://files-api-intb.ciscospark.com/v1/spaces/2c8bb82f-41ff-4c25-815b-d7df95593a32/contents/1b4b0c09-28db-4423-9f81-e3335b7be73e/versions/58ac71fff1814274986836c1c6aa4287/bytes',
        scr: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..GLROpCxPZMcfozcm.1Qd5d9OZYqGqjSGV2COI80u9oqfH4xzlxIcAzC8j2qKhRavkJ2f3S2MVF7D99Y4OTE6o-H1tRJdEYuYkiH7CGtmjQprFLuxt8XX_GO1zALIXAOd28KPrTk5_TGadVzYeNa9ZBsMfLKYPVDwPTGKsGS_nai2TFOOrQB40f2HNVtQA4_Nsf9gN3bMa3rJInD9krUsqEJyG3JoK8ckrXMYQN0fA5uCbzL7zelUV-Hz6bTIZvlbns8fuVTyDL6UiYCVCi5_p8l1rwiY8zD0k10fjQnmkO9XP2nE3R0S_EEg_h-JzrpmN1s-l62V8rH78X1k_sjxH-pS6jgY5nirfV8ZLi9QgyB4mlLxkzU3PlnZewf4fd9OJf1ZWiexJB8Rak-jyEP29VOagofEdzrKuDphwKI6MDYYhAnhMOIybQG26aXamYj40zHXArF8a6sEBThw8EG3OG64Yf6I.aJ4U5eCRK6fPQG11lPcLCg',
        mimeType: 'image/jpeg',
        fileSize: 43984,
        name: 'file.jpg',
        content: [255, 216, 255, 219]
      }
    ]
  },
  {
    activityId: 'f123c910-e695-11e8-98d1-b1349cfbb607',
    actorId: '3ae352c3-f32c-4cf2-8385-8e03c4c801ed',
    actorDisplayName: 'Test User <testuser@email.com>',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/2a69ecbe-a630-4845-9221-818449d7bd7c',
    objectDisplayName: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..qJa0f4tJyQ0aRDBL.9O2t8ddd.YSxOCAUX0BYfRoI0UVwS-g',
    objectId: '',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:13:50.297Z',
    verb: 'post'
  }
];

export default Activities;
