// @webex/internal-plugin-ediscovery/test/unit/spec/activities.js
const Activities = [
  {
    activityId: '029e57c0-e694-11e8-98d1-b1349cfbb607',
    spaceName: 'Test User 2 <testuser2@email.com>',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/19288ee4-d235-4897-93f5-6cdd433ac420',
    objectDisplayName: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..I8dprtpOu6kk4NYO._gE9BUI6E4H4H2FzxStZWq81rJQ7RvEh1idkHtpFQMYC8TYtaN2D_Z053Q.s8mRZ1vME9C7Ne7JmQrgZQ',
    objectId: '',
    targetId: '223b3dbb-0980-34c0-8d4c-52820b6ee5c9',
    timestamp: '2018-11-12T16:00:00.509Z',
    verb: 'post'
  },
  {
    activityId: '14113360-e694-11e8-981c-f5e3fd7b0d6e',
    spaceName: 'Test Space',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    objectId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    objectDisplayName: 'Test User 2 <testuser2@email.com>',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:00:30.102Z',
    verb: 'add'
  },
  {
    activityId: '14115a70-e694-11e8-981c-f5e3fd7b0d6e',
    spaceName: 'Test Space',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    objectId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    objectDisplayName: 'Test User 2 <testuser2@email.com>',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:00:30.102Z',
    verb: 'assignModerator'
  },
  {
    activityId: '14b306c0-e696-11e8-98d1-b1349cfbb607',
    spaceName: 'Test Space',
    actorId: '7765121f-f04e-4bae-8f83-16ac80a3315b',
    actorDisplayName: 'Test User <testuser@email.com>',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/2a69ecbe-a630-4845-9221-818449d7bd7c',
    objectDisplayName: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..bmyKDuR84yTqQVdj.A8IPRqIc3g4.fsgwcN8pdUrHTr4JMf2u8Q',
    objectId: '',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:14:49.958Z',
    verb: 'post'
  },
  {
    activityId: '1b575030-e696-11e8-98d1-b1349cfbb607',
    spaceName: 'Test Space',
    actorId: '66e8b038-3a71-4857-9f42-65a730c93ab2',
    actorDisplayName: 'Test User <testuser@email.com>',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/2a69ecbe-a630-4845-9221-818449d7bd7c',
    objectDisplayName: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..Sg8Q5eItntFtv_6k.-oOiiOc.uAcNMWljkOsRmkvhAjvcHw',
    objectId: '',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:15:01.113Z',
    verb: 'post'
  },
  {
    activityId: '29691160-e694-11e8-981c-f5e3fd7b0d6e',
    spaceName: 'Test Space',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    objectId: '7765121f-f04e-4bae-8f83-16ac80a3315b',
    objectDisplayName: 'Test User 2 <testuser2@email.com>',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:05.910Z',
    verb: 'add'
  },
  {
    activityId: '2b90fa70-e694-11e8-98d1-b1349cfbb607',
    spaceName: 'Test Space',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    objectId: 'a003e2e4-cdfb-4bfb-8a05-8109fe64ca99',
    objectDisplayName: 'Test User 2 <testuser2@email.com>',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:09.527Z',
    verb: 'add'
  },
  {
    activityId: '2dfb1ca0-e694-11e8-981c-f5e3fd7b0d6e',
    spaceName: 'Test Space',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    objectId: '3ae352c3-f32c-4cf2-8385-8e03c4c801ed',
    objectDisplayName: 'Test User 2 <testuser2@email.com>',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:13.578Z',
    verb: 'add'
  },
  {
    activityId: '37ba61b0-e694-11e8-981c-f5e3fd7b0d6e',
    spaceName: 'Test Space',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    objectId: '66e8b038-3a71-4857-9f42-65a730c93ab2',
    objectDisplayName: 'Test User 2 <testuser2@email.com>',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:29.931Z',
    verb: 'add'
  },
  {
    activityId: '3c60b840-e694-11e8-981c-f5e3fd7b0d6e',
    spaceName: 'Test Space',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    objectId: 'cbf0323f-44e9-4b56-8514-a0a7ed3b333a',
    objectDisplayName: 'Test User 2 <testuser2@email.com>',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:37.732Z',
    verb: 'add'
  },
  {
    activityId: '42f411c0-e694-11e8-981c-f5e3fd7b0d6e',
    spaceName: 'Test Space',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    objectId: '91facb42-c96f-441d-b45c-718ba5a71786',
    objectDisplayName: 'Test User 2 <testuser2@email.com>',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:48.764Z',
    verb: 'add'
  },
  {
    activityId: '47a00da0-e694-11e8-98d1-b1349cfbb607',
    spaceName: 'Test Space',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    actorDisplayName: 'Test User <testuser@email.com>',
    objectId: '93b4bd25-ed19-4c72-81a4-a9c8b6241708',
    objectDisplayName: 'Test User 2 <testuser2@email.com>',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:56.602Z',
    verb: 'add'
  },
  {
    activityId: '4aac48e0-e768-11e8-9ad1-cf6f6822e300',
    spaceName: 'Test Space',
    actorId: '66e8b038-3a71-4857-9f42-65a730c93ab2',
    actorDisplayName: 'Test User <testuser@email.com>',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/2a69ecbe-a630-4845-9221-818449d7bd7c',
    objectDisplayName: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..YuNeE_v-jk0XIW7o.nhU.7Ku7utNFVaQ0FQ6qWIanSA',
    objectId: '',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-13T17:19:34.166Z',
    verb: 'post'
  },
  {
    activityId: '7a9ec280-e768-11e8-afdf-b54a3c174c2d',
    spaceName: 'Test Space',
    timestamp: '2018-11-13T17:20:55.464Z',
    verb: 'share',
    actorId: '66e8b038-3a71-4857-9f42-65a730c93ab2',
    actorDisplayName: 'Test User <testuser@email.com>',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/2a69ecbe-a630-4845-9221-818449d7bd7c',
    objectId: '',
    files: [
      {
        displayName: 'file',
        url: 'https://files-api-intb.ciscospark.com/v1/spaces/2c8bb82f-41ff-4c25-815b-d7df95593a32/contents/1b4b0c09-28db-4423-9f81-e3335b7be73e/versions/58ac71fff1814274986836c1c6aa4287/bytes',
        scr: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..GLROpCxPZMcfozcm.1Qd5d9OZYqGqjSGV2COI80u9oqfH4xzlxIcAzC8j2qKhRavkJ2f3S2MVF7D99Y4OTE6o-H1tRJdEYuYkiH7CGtmjQprFLuxt8XX_GO1zALIXAOd28KPrTk5_TGadVzYeNa9ZBsMfLKYPVDwPTGKsGS_nai2TFOOrQB40f2HNVtQA4_Nsf9gN3bMa3rJInD9krUsqEJyG3JoK8ckrXMYQN0fA5uCbzL7zelUV-Hz6bTIZvlbns8fuVTyDL6UiYCVCi5_p8l1rwiY8zD0k10fjQnmkO9XP2nE3R0S_EEg_h-JzrpmN1s-l62V8rH78X1k_sjxH-pS6jgY5nirfV8ZLi9QgyB4mlLxkzU3PlnZewf4fd9OJf1ZWiexJB8Rak-jyEP29VOagofEdzrKuDphwKI6MDYYhAnhMOIybQG26aXamYj40zHXArF8a6sEBThw8EG3OG64Yf6I.aJ4U5eCRK6fPQG11lPcLCg',
        mimeType: 'image/jpeg',
        fileSize: 43984,
        content: [255, 216, 255, 219]
      }
    ]
  },
  {
    activityId: 'f123c910-e695-11e8-98d1-b1349cfbb607',
    spaceName: 'Test Space',
    actorId: '3ae352c3-f32c-4cf2-8385-8e03c4c801ed',
    actorDisplayName: 'Test User <testuser@email.com>',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/2a69ecbe-a630-4845-9221-818449d7bd7c',
    objectDisplayName: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..qJa0f4tJyQ0aRDBL.9O2t8ddd.YSxOCAUX0BYfRoI0UVwS-g',
    objectId: '',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:13:50.297Z',
    verb: 'post'
  },
  { // whiteboard file share
    activityId: '42c268f0-36ad-11e9-b55a-ad84975cb0ba',
    spaceName: 'Test Space',
    timestamp: '2019-02-22T14:22:18.879Z',
    verb: 'share',
    actorId: '91facb42-c96f-441d-b45c-718ba5a71786',
    actorDisplayName: 'Test User <testuser@email.com>',
    targetId: '13b7af30-2b94-11e9-8f7a-59e7a3b4e375',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/0170e243-ce35-4a37-9f5f-aba8906a9c5b',
    files: [
      {
        displayName: 'file',
        url: 'https://files-api-intb.ciscospark.com/v1/spaces/df4d1935-ea38-4627-86e6-c75c1abd756d/contents/616fbd37-e2fc-4c3f-8d6b-b516f128b2bb/versions/19d1293ae9314c9eb4c2c6302815ba77/bytes',
        scr: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..Q5wccnpg7ZTnhP97.gq752Kcnk0TgbqA7bz9fSDnjl16h08ZOcBdrIUk81z26boNNhSm_suQeActKlQ6jt3qpLfEKljAkAtQOGNZekP1G-9dI8qSGxsRBCS04Cccqcn01f4jXOR1OZ9KW5IQariAyycGdwmigRHzSCfodekJz876fDkWXiABBLJUT4ghge3QTOlaJbyPGey8UqfZa2qgbblOcoXXT4CCrkznBeRU6qxisMoZmBySgXV0RkXX4IwwP2EoItcgzFce4Ewzf0xp1dG1uf9lAKX1u-wRxeFCu_Lu0uDRvFRLqnYgVXSYPLaFmJ4S7txBZES6b3Mal6f7QnYkyoRwKc-iN28-r6emFiWcn8xSu34PQdF_zAt17obxHmi7aKPwrsJWlHatHcWv_P4K4syWh6-9S4BYDwMZ0STcreI4TKUeQuP_SdZkONLWLMyN43hd9NzLLsR5jEYamaIifLkg.BSdIDNLDdZPh8zNEOaKlBw',
        mimeType: 'image/png',
        fileSize: 77964,
        content: [255, 216, 255, 219]
      }
    ]
  },
  { // whiteboard snapshot
    activityId: '6a580050-3cc2-4691-9e10-f4c92de40c00',
    spaceName: 'Test Space',
    timestamp: '2019-02-22T14:24:06.159Z',
    verb: 'share',
    actorId: '00000000-0000-0000-0000-000000000000',
    actorDisplayName: '',
    targetId: '13b7af30-2b94-11e9-8f7a-59e7a3b4e375',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/23501ab1-05c0-4f29-a95f-cce2a6f45ad7',
    files: [{
      displayName: 'Whiteboard Activity',
      url: 'https://files-api-intb.ciscospark.com/v1/spaces/9721b65f-1035-48b5-9d8e-d8975be934a7/contents/b6f8b746-19bb-48b1-9c3e-b253ebcf9f57/versions/c241f9ba975549318e4693387f0ccd38/bytes',
      scr: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..Rkr62sA0UnwxehCW.11IcP1vE9b7zZ-MzmYje2GlLMBQIVkHhpJZi24VQNE4N9pQG1-esyxfJX7_uFNkqAo8D4Pj0n21PaKlpt6F1x4hbpeGUya5j2BSeDoQp9OvwlkQPauntuu8Nzb_BsFqEv0g4gJ7ysBpgZWFzjRtFw7SHwwR7fK97qc1BH7hHPc6coPzdt9RaAhgtzqJV1rdXfuEm01FrB2biwL52CuFl0BoLRm5Z2HfpH0_mUTSwdDOWdb9--NzEzcjIJ-lnIgvgrimoWbSuxMRbQd0AM1ALMsDNfOpx9R1zxyw9mwBf6lZQAdAiyLBmePgTHAetnshwevmTJ4xvWr6dJ4fd7eysiE6JjXYnMG2J18KKLGqtNyYMImbGHsNlPDB6b7t5aeN1EPbyJFvblg1YjqawwIqMDTfLeX7HVZu4MDXQkt7O6bLopG51oyBtysQmVSH3GlRnGlG21oYEe8Q.Cg3Xu5ozYSnFEJLh5mSaBA',
      mimeType: 'image/png',
      fileSize: 75057,
      content: [255, 216, 255, 219]
    }]
  }
];

export default Activities;
