// @webex/internal-plugin-voicea/src/constants.ts
export const EVENT_TRIGGERS = {
  VOICEA_ANNOUNCEMENT: 'meeting:voicea:announcement',
  CAPTION_LANGUAGE_UPDATE: 'meeting:voicea:captionLanguageUpdate',
  SPOKEN_LANGUAGE_UPDATE: 'meeting:voicea:spokenLanguageUpdate',
  CAPTIONS_TURNED_ON: 'meeting:voicea:captionOn',
  TRANSCRIBING_ON: 'meeting:voicea:transcribingOn',
  TRANSCRIBING_OFF: 'meeting:voicea:transcribingOff',

  NEW_CAPTION: 'meeting:voicea:newCaption',
  EVA_COMMAND: 'meeting:voicea:wxa',
  HIGHLIGHT_CREATED: 'meeting:voicea:highlightCreated'
};

export const VOICEA_RELAY_TYPES = {
  ANNOUNCEMENT: 'voicea.annc',
  CLIENT_ANNOUNCEMENT: 'client.annc',
  TRANSLATION_REQUEST: 'voicea.transl.req',
  TRANSLATION_RESPONSE: 'voicea.transl.rsp',
  TRANSCRIPTION: 'voicea.transcription'
};

export const TRANSCRIPTION_TYPE = {
  UNKNOWN: 'unknown',
  EVA_WAKE: 'eva_wake',
  EVA_THANKS: 'eva_thanks',
  EVA_CANCEL: 'eva_cancel',
  HIGHLIGHT_CREATED: 'highlight_created',
  TRANSCRIPT_INTERIM_RESULTS: 'transcript_interim_results',
  TRANSCRIPT_FINAL_RESULT: 'transcript_final_result'

};
