// @webex/plugin-meetings/src/stats/history.js
/**
 * stores the last x filtered stream datas
 */
export default class StatsHistory {
  /**
   * instantiate our wrapped history array
   * @param {Number} max
   */
  constructor(max) {
    this.history = [];
    this.max = max;
  }

  /**
   * @returns {Array} the array of stats reports, read from [0] = most recent to [length - 1] = least recent
   */
  get() {
    return this.history;
  }

  /**
   * deletes the history array and resets it
   * @returns {undefined}
   */
  clearHistory() {
    this.history = [];
  }

  /**
   * gets the stored stat
   * @param {Number} index the location
   * @returns {Object} the stat at location index
   */
  getHistoryAt(index) {
    return this.history[index];
  }

  /**
   * gets the most recently stored stat
   * @returns {Object} the most recently added stat to the history recorder
   */
  getMostRecent() {
    return this.history.length > 1 ? this.history[1] : undefined;
  }

  /**
   * adds a history entry into the array at the head, removes from the tail
   * if too large, returns the old tail if removed
   * @param {WebRTCData} data filtered stats report to add to the history array
   * @returns {Object} the removed stats report at the end that had to be removed
   * for the new stats data report to be added to the front
   */
  addHistory(data) {
    let removed;

    if (this.history.length >= this.max) {
      removed = this.history.pop();
    }
    this.history.unshift(data);

    return removed;
  }
}
