// @webex/plugin-meetings/src/stats/index.js
import {StatelessSparkPlugin} from '@ciscospark/spark-core';

import {MEETINGS} from '../constants';
import MeetingStats from '../stats/stats';

/**
 * @class WebRTCStats
 */
export default class WebRTCStats extends StatelessSparkPlugin {
    namespace = MEETINGS;

  /**
    *
    * @param {Object} attrs
    * @param {Object} options
    * @param {Object} optionalCreateOptions - optional {media: {}, screen: {}, see MeetingStats options}
   */
    constructor(attrs, options, optionalCreateOptions) {
      super({}, options);
      this.attrs = attrs;
      this.options = options;
      this.media = undefined;
      if (optionalCreateOptions && optionalCreateOptions.media) {
        this.withMedia(optionalCreateOptions.media);
      }
    }

  /**
   * @returns {WebRTCStats}
   */
    build() {
      return this;
    }

  /**
   * constructs a MeetingStats object on this instance for media stream
   * @param {Object} options - optional, see MeetingStats options
   * @returns {WebRTCStats}
   */
    withMedia(options) {
      this.media = new MeetingStats(this.attrs, this.options, options);
      return this;
    }

  /**
   * @param {MeetingStats} media
   * @returns {undefined}
   */
    setMedia(media) {
      this.media = media;
    }

  /**
   * @returns {MeetingStats}
   */
    getMedia() {
      return this.media;
    }
}
