// @webex/plugin-meetings/src/stats/index.js
import uuid from 'uuid';

import MeetingStats from '../stats/stats';

/**
 * @class WebRTCStats
 */
export default class WebRTCStats {

  /**
   *
   */
  constructor() {
    this.media = new MeetingStats();
    this.screen = new MeetingStats();
    this.id = undefined;
  }

  /**
   * constructs an id to match this stats object
   * takes params as precedence
   * @param {String} id
   * @returns {WebRTCStats}
   */
  withId(id) {
    if (id) {
      this.setId(id);
      return this;
    }
    const guid = uuid.v4();
    this.setId(guid);
    return this;
  }

  /**
   * @returns {WebRTCStats}
   */
  build() {
    return this;
  }

  /**
   * constructs a MeetingStats object on this instance for media stream
   * @returns {WebRTCStats}
   */
  withMedia() {
    this.media = new MeetingStats();
    return this;
  }

  /**
   * constructs a interval object on this instance
   * @returns {WebRTCStats}
   */
  withScreen() {
    this.screen = new MeetingStats();
    return this;
  }

  /**
   * @param {String} id
   * @returns {undefined}
   */
  setId(id) {
    this.id = id;
  }

  /**
   * @param {MeetingStats} media
   * @returns {undefined}
   */
  setMedia(media) {
    this.media = media;
  }

  /**
   * @param {MeetingStats} screen
   * @returns {undefined}
   */
  setScreen(screen) {
    this.screen = screen;
  }

  /**
   * @returns {String}
   */
  getId() {
    return this.id;
  }

  /**
   * @returns {MeetingStats}
   */
  getMedia() {
    return this.media;
  }

  /**
   * @returns {MeetingStats}
   */
  getScreen() {
    return this.screen;
  }
}
