// @webex/plugin-meetings/test/integration/spec/sparkTestUsers.js

const testUser = require('@ciscospark/test-helper-test-users');
const Spark = require('@webex/webex-core').default;

const config = require('./sparkconfig.js');


// Include the plugins which you feel will be used by the spark instance
require('@webex/internal-plugin-mercury');
require('@webex/internal-plugin-user');
require('@webex/internal-plugin-wdm');
require('@webex/internal-plugin-conversation');
require('@webex/plugin-meetings');

const generateTestUsers = (options) => testUser.create({count: options.count})
  .then((userSet) => {
    const mercuryPromise = [];
    const devicePromise = [];

    if (userSet.length === 0) {
      return Promise.reject(new Error('Test user not created'));
    }

    userSet.forEach((user) => {
      user.spark = new Spark({
        config: config.spark,
        credentials: {
          authorization: user.token
        }
      });

      console.log(user.spark);
      mercuryPromise.push(user.spark.internal.mercury.connect());
      devicePromise.push(user.spark.internal.device.register());
    });

    return Promise.all(mercuryPromise)
      .then(() => Promise.all(devicePromise))
      .then(() => userSet)
      .catch((error) => {
        console.log(error);
      });
  })
  .catch((error) => {
    console.log('ERROR', error);
  });

const remove = (userSet) => testUser.remove(userSet);

const reserveCMR = (user) => user.spark.request({
  method: 'POST',
  uri: 'https://whistler-prod.onint.ciscospark.com/api/v1/reservations',
  headers: {
    authorization: `Bearer ${user.spark.credentials.supertoken.access_token}`,
    'cisco-no-http-redirect': null,
    'spark-user-agent': null,
    trackingid: `ITCLIENT_ ${user.id} _0_imi:true`
  },
  body: {
    resourceType: 'CMR_3',
    reservedBy: 'SDK_TEST_USER',
    requestMetaData: {
      emailAddress: user.emailAddress,
      loginType: 'loginGuest'
    }
  }
})
  .then((res) => {
    console.log('SUCCESS ', res);
  });

module.exports = {
  generateTestUsers,
  remove,
  reserveCMR
};
