// @ciscospark/plugin-phone/src/config.js
/*!
 * Copyright (c) 2015-2017 Cisco Systems, Inc. See LICENSE file.
 */

export default {
  phone: {
    /**
     * @type {number}
     * @memberof config.phone
     */
    audioBandwidthLimit: 64000,
    /**
     * @type {number}
     * @memberof config.phone
     */
    videoBandwidthLimit: 1000000,
    /**
     * Group calling support should work, but necessarily introduces some
     * breaking changes to the API. In order to get feedback, we're keeping the
     * API unchanged by default and requiring folks to opt-in to these
     * breakages.
     * @type {boolean}
     * @memberof config.phone
     */
    enableExperimentalGroupCallingSupport: false,
    hangupIfLastActive: {
      /**
       * Indicates if a call (i.e. two-party audio-video experience) should be
       * ended automatically if the current user is the last active member of
       * the experience
       * @type {boolean}
       * @memberof config.phone.hangupeIfLastActive
       */
      call: true,
      /**
       * Indicates if a meeting (i.e. multi-party audio-video experience) should
       * be ended automatically if the current user is the last active member of
       * the experience
       * @type {boolean}
       * @memberof config.phone.hangupeIfLastActive
       */
      meeting: false
    }
  }
};
