// @ciscospark/internal-plugin-ediscovery/test/unit/spec/activities.js
const Activities = [
  {
    activityId: '029e57c0-e694-11e8-98d1-b1349cfbb607',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/19288ee4-d235-4897-93f5-6cdd433ac420',
    objectDisplayName: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..I8dprtpOu6kk4NYO._gE9BUI6E4H4H2FzxStZWq81rJQ7RvEh1idkHtpFQMYC8TYtaN2D_Z053Q.s8mRZ1vME9C7Ne7JmQrgZQ',
    objectId: '',
    targetId: '223b3dbb-0980-34c0-8d4c-52820b6ee5c9',
    timestamp: '2018-11-12T16:00:00.509Z',
    verb: 'post'
  },
  {
    activityId: '14113360-e694-11e8-981c-f5e3fd7b0d6e',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    objectId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:00:30.102Z',
    verb: 'add'
  },
  {
    activityId: '14115a70-e694-11e8-981c-f5e3fd7b0d6e',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    objectId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:00:30.102Z',
    verb: 'assignModerator'
  },
  {
    activityId: '14b306c0-e696-11e8-98d1-b1349cfbb607',
    actorId: '7765121f-f04e-4bae-8f83-16ac80a3315b',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/2a69ecbe-a630-4845-9221-818449d7bd7c',
    objectDisplayName: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..bmyKDuR84yTqQVdj.A8IPRqIc3g4.fsgwcN8pdUrHTr4JMf2u8Q',
    objectId: '',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:14:49.958Z',
    verb: 'post'
  },
  {
    activityId: '1b575030-e696-11e8-98d1-b1349cfbb607',
    actorId: '66e8b038-3a71-4857-9f42-65a730c93ab2',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/2a69ecbe-a630-4845-9221-818449d7bd7c',
    objectDisplayName: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..Sg8Q5eItntFtv_6k.-oOiiOc.uAcNMWljkOsRmkvhAjvcHw',
    objectId: '',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:15:01.113Z',
    verb: 'post'
  },
  {
    activityId: '29691160-e694-11e8-981c-f5e3fd7b0d6e',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    objectId: '7765121f-f04e-4bae-8f83-16ac80a3315b',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:05.910Z',
    verb: 'add'
  },
  {
    activityId: '2b90fa70-e694-11e8-98d1-b1349cfbb607',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    objectId: 'a003e2e4-cdfb-4bfb-8a05-8109fe64ca99',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:09.527Z',
    verb: 'add'
  },
  {
    activityId: '2dfb1ca0-e694-11e8-981c-f5e3fd7b0d6e',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    objectId: '3ae352c3-f32c-4cf2-8385-8e03c4c801ed',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:13.578Z',
    verb: 'add'
  },
  {
    activityId: '37ba61b0-e694-11e8-981c-f5e3fd7b0d6e',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    objectId: '66e8b038-3a71-4857-9f42-65a730c93ab2',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:29.931Z',
    verb: 'add'
  },
  {
    activityId: '3c60b840-e694-11e8-981c-f5e3fd7b0d6e',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    objectId: 'cbf0323f-44e9-4b56-8514-a0a7ed3b333a',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:37.732Z',
    verb: 'add'
  },
  {
    activityId: '42f411c0-e694-11e8-981c-f5e3fd7b0d6e',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    objectId: '91facb42-c96f-441d-b45c-718ba5a71786',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:48.764Z',
    verb: 'add'
  },
  {
    activityId: '47a00da0-e694-11e8-98d1-b1349cfbb607',
    actorId: 'e0325f64-ae51-4821-9f8b-9d238f49d81b',
    objectId: '93b4bd25-ed19-4c72-81a4-a9c8b6241708',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:01:56.602Z',
    verb: 'add'
  },
  {
    activityId: '4aac48e0-e768-11e8-9ad1-cf6f6822e300',
    actorId: '66e8b038-3a71-4857-9f42-65a730c93ab2',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/2a69ecbe-a630-4845-9221-818449d7bd7c',
    objectDisplayName: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..YuNeE_v-jk0XIW7o.nhU.7Ku7utNFVaQ0FQ6qWIanSA',
    objectId: '',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-13T17:19:34.166Z',
    verb: 'post'
  },
  {
    activityId: '7a9ec280-e768-11e8-afdf-b54a3c174c2d',
    actorId: '66e8b038-3a71-4857-9f42-65a730c93ab2',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/2a69ecbe-a630-4845-9221-818449d7bd7c',
    objectId: '',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-13T17:20:55.464Z',
    verb: 'share'
  },
  {
    activityId: 'f123c910-e695-11e8-98d1-b1349cfbb607',
    actorId: '3ae352c3-f32c-4cf2-8385-8e03c4c801ed',
    encryptionKeyUrl: 'kms://ciscospark.com/keys/2a69ecbe-a630-4845-9221-818449d7bd7c',
    objectDisplayName: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..qJa0f4tJyQ0aRDBL.9O2t8ddd.YSxOCAUX0BYfRoI0UVwS-g',
    objectId: '',
    targetId: '14110c50-e694-11e8-981c-f5e3fd7b0d6e',
    timestamp: '2018-11-12T16:13:50.297Z',
    verb: 'post'
  }
];

export default Activities;
