// @webex/common/src/constants.js
export const WEBEX_EVENT = {
  TEAMS_ACTIVITY: 'event:conversation.activity'
};

export const WEBEX_SPACE_TYPE = {
  DIRECT: 'direct',
  GROUP: 'group'
};

export const API_ACTIVITY_FIELD = {
  ACTOR: 'actor',
  OBJECT: 'object',
  TARGET: 'target'
};

export const API_ACTIVITY_VERB = {
  ACKNOWLEDGE: 'acknowledge',
  POST: 'post',
  SHARE: 'share',
  DELETE: 'delete',
  ADD: 'add',
  LEAVE: 'leave',
  ADD_MODERATOR: 'assignModerator',
  REMOVE_MODERATOR: 'unassignModerator',
  HIDE: 'hide'
};

export const SDK_EVENT = {
  CREATED: 'created',
  DELETED: 'deleted',
  UPDATED: 'updated',
  SEEN: 'seen'
};

export const SDK_EVENT_OWNER = {
  CREATOR: 'creator',
  ORG: 'org'
};

export const SDK_EVENT_STATUS = {
  ACTIVE: 'active',
  DISABLED: 'disabled'
};

export const SDK_EVENT_RESOURCE = {
  MEMBERSHIPS: 'memberships',
  MESSAGES: 'messages'
};

export const hydraTypes = {
  CONTENT: 'CONTENT',
  MEMBERSHIP: 'MEMBERSHIP',
  MESSAGE: 'MESSAGE',
  ORGANIZATION: 'ORGANIZATION',
  PEOPLE: 'PEOPLE',
  ROOM: 'ROOM',
  TEAM: 'TEAM'
};
