// @ciscospark/internal-plugin-wdm/src/device/device.js
/*!
 * Copyright (c) 2015-2017 Cisco Systems, Inc. See LICENSE file.
 */

import {oneFlight, tap} from '@ciscospark/common';
import {safeSetTimeout} from '@ciscospark/common-timers';
import {omit} from 'lodash';
import util from 'util';
import FeaturesModel from './features-model';
import ServiceCollection from './service-collection';
import {persist, waitForValue, SparkPlugin} from '@ciscospark/spark-core';
import * as Url from 'url';

/**
 * Decides if this device should be persisted to boundedStorage, based on
 * this.config.ephemeral.
 * @returns {Boolean}
 */
function decider() {
  return !this.config.ephemeral;
}

const Device = SparkPlugin.extend({
  children: {
    features: FeaturesModel
  },

  collections: {
    serviceCatalog: ServiceCollection
  },

  idAttribute: `url`,

  namespace: `Device`,

  props: {
    // deviceType doesn't have any real value, but we need to send it during
    // device refresh to make sure we don't get back an ios device url
    deviceType: `string`,
    intranetInactivityDuration: `number`,
    intranetInactivityCheckUrl: `string`,
    modificationTime: `string`,
    searchEncryptionKeyUrl: `string`,
    services: {
      // Even though @jodykstr will tell you the docs claim you don't need to
      // initialize `object` properties, the docs lie.
      default() {
        return {};
      },
      type: `object`
    },
    serviceHostMap: {
      default() {
        return {
          serviceLinks: {},
          hostCatalog: {}
        };
      },
      type: `object`
    },
    url: `string`,
    userId: `string`,
    /**
     * Notifies the client if file sharing is disabled.
     * Currently, the values for it are:
     * - BLOCK_BOTH
     * - BLOCK_UPLOAD
     * @instance
     * @memberof Device
     * @type {string}
     */
    webFileShareControl: `string`,
    webSocketUrl: `string`
  },

  derived: {
    registered: {
      deps: [`url`],
      fn() {
        return Boolean(this.url);
      }
    },
    currentWebSocketUrl: {
      deps: [`webSocketUrl`, `serviceHostMap`, `serviceCatalog`],
      fn() {
        if (this.webSocketUrl) {
          return this._useUrlFromServiceCatalog(this.webSocketUrl, `mercuryConnection`);
        }
        return null;
      }
    }
  },

  session: {
    // Fun Fact: setTimeout returns a Timer object instead of a Number in Node 6
    // or later
    logoutTimer: `any`,
    lastUserActivityDate: `number`
  },

  @waitForValue(`@`)
  determineService(url) {
    for (const key of Object.keys(this.services)) {
      const serviceUrl = this.services[key];
      if (url.startsWith(serviceUrl)) {
        // "ServiceUrl" is 10 characters
        return Promise.resolve(key.substr(0, key.length - 10));
      }
    }

    return Promise.reject(new Error(`${url} does not reflect a known service`));
  },

  @waitForValue(`@`)
  getServiceUrl(service) {
    return this._getServiceUrl(this.services, service)
      .then((isServiceUrl) => isServiceUrl || this.getPreDiscoveryServiceUrl(service));
  },

  getPreDiscoveryServiceUrl(service) {
    // The Promise.resolve here is temporary. A future PR will make the
    // corresponding _ method async to allow for lazy device registration
    return Promise.resolve(this._getServiceUrl(this.config.preDiscoveryServices, service));
  },

  markUrlFailedAndGetNew(service) {
    if (!service) {
      return Promise.reject(new Error(`\`service\` is a required parameter`));
    }

    const s = this.serviceCatalog.get(`${service}ServiceUrl`);
    if (s) {
      s.markHostFailed();
      s.cycleHost();
      const host = s.getCurrentHost();
      if (host.failed) {
        this.logger.info(`device: all known hosts have failed, refreshing for new ones`);
        // we've cycled though all hosts, now we rely on wdm service to return
        // new available datacenters
        return new Promise((resolve) => {
          this.refresh();
          this.on(`serviceCatalogUpdated`, () => {
            resolve(s.url);
          });
        });
      }

      return Promise.resolve(s.url)
        // this will cause currentWebSocketUrl to use the new host
        .then(tap(() => this.trigger(`change:serviceCatalog`)));
    }

    return Promise.reject(new Error(`service ${service} not found`));
  },

  @persist(`@`, decider)
  initialize(...args) {
    Reflect.apply(SparkPlugin.prototype.initialize, this, args);

    // Propagate change(:[attribute]) events from collections
    [`developer`, `entitlement`, `user`].forEach((collectionName) => {
      this.features.on(`change:${collectionName}`, (model, value, options) => {
        this.trigger(`change`, this, options);
        this.trigger(`change:features`, this, this.features, options);
      });
    });

    this.on(`change:serviceHostMap`, this._updateServiceCatalog);

    this.listenToAndRun(this, `change:intranetInactivityCheckUrl`, () => this._resetLogoutTimer());
    this.listenToAndRun(this, `change:intranetInactivityDuration`, () => this._resetLogoutTimer());
    this.listenTo(this.spark, `user-activity`, () => {this.lastUserActivityDate = Date.now();});
  },

  /**
   * Don't log the features object
   * @param {number} depth
   * @returns {Object}
   */
  inspect(depth) {
    return util.inspect(omit(this.serialize(), `features`), {depth});
  },

  isPreDiscoveryService(service) {
    // The Promise.resolve here is temporary. A future PR will make the
    // corresponding _ method async to allow for lazy device registration
    return Promise.resolve(this._isService(this.config.preDiscoveryServices, service));
  },

  isPreDiscoveryServiceUrl(uri) {
    // The Promise.resolve here is temporary. A future PR will make the
    // corresponding _ method async to allow for lazy device registration
    return Promise.resolve(this._isServiceUrl(this.config.preDiscoveryServices, uri));
  },

  @waitForValue(`@`)
  isService(service) {
    return this._isService(this.services, service)
      .then((_isService) => _isService || this.isPreDiscoveryService(service));
  },

  @waitForValue(`@`)
  isServiceUrl(uri) {
    // The Promise.resolve here is temporary. A future PR will make the
    // corresponding _ method async to allow for lazy device registration
    return Promise.resolve(this._isServiceUrl(this.services, uri));
  },

  @waitForValue(`@`)
  isSpecificService(service, key) {
    if (key === service) {
      return Promise.resolve(true);
    }

    return this.getServiceUrl(service)
      .then((serviceUrl) => key.includes(serviceUrl));
  },

  _getServiceUrl(target, service) {
    /* istanbul ignore if */
    if (!target) {
      return Promise.reject(new Error(`\`target\` is a required parameter`));
    }

    if (!service) {
      return Promise.reject(new Error(`\`service\` is a required parameter`));
    }

    const s = this.serviceCatalog.get(`${service}ServiceUrl`);
    if (s) {
      return Promise.resolve(s.url);
    }

    return Promise.resolve(target[`${service}ServiceUrl`]);
  },

  _isService(target, service) {
    return this._getServiceUrl(target, service)
      .then((url) => Boolean(url));
  },

  _isServiceUrl(target, uri) {
    if (!uri) {
      return Promise.reject(new Error(`\`uri\` is a required parameter`));
    }

    for (const key in target) {
      if (target[key] && uri.indexOf(target[key]) === 0) {
        return Promise.resolve(true);
      }
    }

    return Promise.resolve(false);
  },

  @oneFlight
  @waitForValue(`@`)
  refresh() {
    this.logger.info(`device: refreshing`);

    if (!this.registered) {
      this.logger.info(`device: device not registered, registering`);
      return this.register();
    }

    const body = omit(this.serialize(), `features`, `mediaClusters`);
    if (this.config.ephemeral) {
      body.ttl = this.config.ephemeralDeviceTTL;
    }

    return this.request({
      method: `PUT`,
      uri: this.url,
      body
    })
      .then((res) => this._processRegistrationSuccess(res))
      .catch((reason) => {
        if (reason.statusCode === 404) {
          // If we get a 404, it means the device is no longer valid and we need
          // to register a new one.
          this.logger.info(`device: refresh failed with 404, attempting to register new device`);
          this.clear();
          return this.register();
        }
        return Promise.reject(reason);
      });
  },

  @oneFlight
  @waitForValue(`@`)
  register() {
    /* eslint no-invalid-this: [0] */
    this.logger.info(`device: registering`);

    if (this.registered) {
      this.logger.info(`device: device already registered, refreshing`);
      return this.refresh();
    }

    const body = this.config.defaults;
    if (this.config.ephemeral) {
      body.ttl = this.config.ephemeralDeviceTTL;
    }

    return this.request({
      method: `POST`,
      service: `wdm`,
      resource: `devices`,
      body
    })
      .then((res) => this._processRegistrationSuccess(res));
  },

  @oneFlight
  @waitForValue(`@`)
  unregister() {
    this.logger.info(`device: unregistering`);

    if (!this.url) {
      this.logger.warn(`device: not registered`);
      return Promise.resolve();
    }

    return this.request({
      uri: this.url,
      method: `DELETE`
    })
      .then(() => this.clear());
  },

  clear(...args) {
    clearTimeout(this.refreshTimer);
    Reflect.apply(SparkPlugin.prototype.clear, this, args);
  },

  _useUrlFromServiceCatalog(uri, service) {
    const s = this.serviceCatalog.get(`${service}ServiceUrl`);
    if (!s) {
      return uri;
    }
    return s.replaceUrlWithCurrentHost(uri);
  },

  _processRegistrationSuccess(res) {
    this.logger.info(`device: received registration payload`);
    this.set(res.body);
    if (this.config.ephemeral) {
      this.logger.info(`device: enqueing device refresh`);
      const delay = (this.config.ephemeralDeviceTTL / 2 + 60) * 1000;
      this.refreshTimer = safeSetTimeout(() => this.refresh(), delay);
    }
  },

  _updateServiceCatalog(newRegistration) {
    Object.keys(newRegistration.services).forEach((service) => {
      const uri = newRegistration.services[service];
      const u = Url.parse(uri);
      const hosts = newRegistration.serviceHostMap.hostCatalog[u.host];
      const existingService = this.serviceCatalog.get(service);
      if (existingService) {
        existingService.defaultUrl = uri;
        existingService.availableHosts = hosts || [];
      }
      else {
        this.serviceCatalog.add({
          service,
          defaultUrl: uri,
          availableHosts: hosts || []
        });
      }
    });
    this.trigger(`serviceCatalogUpdated`);
  },

  _resetLogoutTimer() {
    clearTimeout(this.logoutTimer);
    this.unset(`logoutTimer`);
    if (this.config.enableInactivityEnforcement && this.intranetInactivityCheckUrl && this.intranetInactivityDuration) {
      this.once(`change:lastUserActivityDate`, () => this._resetLogoutTimer());

      const timer = safeSetTimeout(() => {
        this.spark.request({
          headers: {
            'cisco-no-http-redirect': null,
            'spark-user-agent': null,
            trackingid: null
          },
          method: `GET`,
          uri: this.intranetInactivityCheckUrl
        })
          .catch(() => {
            this.logger.info(`device: did not reach internal ping endpoint; logging out after inactivity on a public network`);
            return this.spark.logout();
          })
          .catch((reason) => {
            this.logger.warn(`device: logout failed`, reason);
          });
      }, this.intranetInactivityDuration * 1000);

      this.logoutTimer = timer;
    }
  }
});

export default Device;
