// @webex/plugin-meetings/src/devices/index.js
/*!
 * Copyright (c) 2015-2018 Cisco Systems, Inc. See LICENSE file.
 */

/**
 *
 *
 * @export
 * @class Devices
 */
export default class Devices {
  /**
   * Creates an instance of Devices.
   * @param {*} {
   *     id, status, isPaired, pairedToken, type
   *   }
   * @memberof Devices
   */
  constructor({
    id, status, isPaired, pairedToken, type
  }) {
    this.id = id;
    this.status = status;
    this.isPaired = isPaired;
    this.pairedToken = pairedToken;
    this.type = type;
  }

  /**
   *
   * Request pairing of a device, AKA resource.
   * @param {string} id
   * @param {string} token
   * @returns {Promise}
   * @memberof Devices
   */
  requestPair(id, token) {
    return new Promise((resolve, reject) => {
      this.id = id;
      this.pairedToken = token;
      resolve(true);
    });
  }

  /**
   * TODO: Can a participant pair with a single device or multiple devices?
   *
   * @param {string} pin
   * @returns {Promise}
   * @memberof Devices
   */
  pair(pin) {
    return new Promise((resolve, reject) => {
      // do something with the pin/challenge
      this.isPaired = true;
      resolve(true);
    });
  }

  /**
   *
   * Requires someone to be paired.
   * @param {string} id
   * @returns {Promise}
   * @memberof Devices
   */
  join(id) {
    return new Promise((resolve, reject) => {
      if (!this.isPaired) reject();
      this.id = id;
    });
  }

  /**
   *
   * TODO: is this possible?
   * @returns {Promise}
   * @memberof Devices
   */
  muteVideo() {
    return new Promise((resolve, reject) => {

    });
  }

  /**
   *
   * TODO: is this possible?
   * @returns {Promise}
   * @memberof Devices
   */
  unmuteVideo() {
    return new Promise((resolve, reject) => {

    });
  }

  /**
   *
   * This may not be automatically returned from all services depending on the scenario.
   * We may need to query lyra to get the status of the device, for example.
   * @returns {Promise}
   * @memberof Devices
   */
  muteAudio() {
    return new Promise((resolve, reject) => {

    });
  }

  /**
   *
   *
   * @returns {Promise}
   * @memberof Devices
   */
  unpair() {
    return new Promise((resolve, reject) => {

    });
  }

  /**
   *
   * Requires someone to be paired.
   * @returns {Promise}
   * @memberof Devices
   */
  leave() {
    return new Promise((resolve, reject) => {
      if (!this.isPaired) reject();
    });
  }
}
