// @webex/plugin-meetings/src/MEMBERSLALA/member.js
/*!
 * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
 */


import {MEETINGS} from '../constants';

/**
 * @class Member
 */
export default class Member {
  namespace = MEETINGS;

  /**
   * @param {Object} participant - the locus participant
   * @param {Object} options - optional constructor params
   * @returns {Member}
   * @memberof Member
   * shallow copies locus participant object into a meeting member object
   */
  constructor(participant, options = {}) {
    this.isAudioMuted = null;
    this.isVideoMuted = null;
    this.id = null;
    this.isSelf = null;
    this.isHost = null;
    this.isInLobby = null;
    this.isInMeeting = null;
    this.isNotInMeeting = null;
    this.processParticipant(participant);
    this.processOptions(options);
  }

  processParticipant(participant) {
    this.participant = participant;
    this.id = participant.id;
    this.isAudioMuted = participant.muted; // TODO:
    this.isVideoMuted = participant.muted; // TODO:
  }

  processOptions(options) {
    this.isSelf = options.isSelf;
    this.isHost = options.isHost;
    this.isInLobby = options.isInLobby;
    this.isInMeeting = options.isInMeeting;
    this.isNotInMeeting = options.isNotInMeeting;
  }
}
