// @webex/plugin-meetings/src/common/logs/logger-proxy.js
import LoggerConfig from './logger-config';

const LoggerProxy = {
  logger: {
    group: () => { console.error('LoggerProxy->group#NO LOGGER DEFINED'); },
    groupEnd: () => { console.error('LoggerProxy->groupEnd#NO LOGGER DEFINED'); },
    info: () => { console.error('LoggerProxy->info#NO LOGGER DEFINED'); },
    log: () => { console.error('LoggerProxy->log#NO LOGGER DEFINED'); },
    error: () => { console.error('LoggerProxy->error#NO LOGGER DEFINED'); },
    warn: () => { console.error('LoggerProxy->warn#NO LOGGER DEFINED'); },
    trace: () => { console.error('LoggerProxy->trace#NO LOGGER DEFINED'); },
    debug: () => { console.error('LoggerProxy->debug#NO LOGGER DEFINED'); }
  }
};

LoggerProxy.set = (logger) => {
  if (!LoggerConfig.enable) {
    LoggerProxy.logger = {
      group: () => {},
      groupEnd: () => {},
      info: () => {},
      log: () => {},
      error: () => {},
      warn: () => {},
      trace: () => {},
      debug: () => {}
    };
  }
  LoggerProxy.logger = logger;
};

LoggerProxy.get = () => LoggerProxy.logger;

export default LoggerProxy;
