// @webex/plugin-meetings/src/media/screen-share.js
/* global chrome */

import bowser from 'bowser';

const chromeShare = {
  EXTENSION_INSTALLED: false,
  REQUEST_ID: null,
  STREAM_ID: null,
  extensionId: 'ifbdadgbpalmagalacllfaflfakmfkac',
  WEB_STORE_URL: 'https://chrome.google.com/webstore/detail/ifbdadgbpalmagalacllfaflfakmfkac',


  // Will return `ping` as false if the extension is not found, Else will retuen the source id for the stream
  verifyExtension() {
    return new Promise(((resolve, reject) => {
      if (!bowser.chrome) {
        resolve();
      }
      chrome.runtime.sendMessage(chromeShare.extensionId, {type: 'ping'},
        (response) => {
          if (response && response.type === 'ping' && response.success) {
            chromeShare.EXTENSION_INSTALLED = true;
            resolve(response);
          }
          else {
            console.error("chrome extension 'doesn't' exist");
            // eslint-disable-next-line prefer-promise-reject-errors
            reject({ping: false});
          }
        });
    }));
  },
  installExtension() {
    return new Promise(((resolve, reject) => {
      chrome.webstore.install(chromeShare.WEB_STORE_URL,
        () => {
          chromeShare.EXTENSION_INSTALLED = true;
          resolve();
        },
        (e) => {
          console.error(`Chrome extension install error = ${e}`);
          reject();
        });
    }));
  },
  requestShare(options) {
    return new Promise(((resolve, reject) => {
      chrome.runtime.sendMessage(chromeShare.extensionId, {type: 'start', options},
        (response) => {
          if (response && response.success && response.sourceId) {
            resolve(response);
          }
          else {
            reject();
          }
        });
    }));
  }
};

export default chromeShare;
