// @webex/plugin-phone/browsers.js
/*!
 * Copyright (c) 2015-2020 Cisco Systems, Inc. See LICENSE file.
 */

// eslint-disable-next-line strict


/* eslint camelcase: [0] */
module.exports = function createBrowsers() {
  if (process.env.SC_TUNNEL_IDENTIFIER || process.env.CI || process.env.CIRCLECI || process.env.SAUCE) {
    return {
      sl_chrome_latest_macOS_High_Sierra: {
        base: 'SauceLabs',
        platform: 'macOS 10.13',
        browserName: 'Chrome',
        version: 'latest',
        extendedDebugging: true,
        'goog:chromeOptions': {
          args: [
            '--use-fake-device-for-media-stream',
            '--use-fake-ui-for-media-stream'
          ]
        }
      },
      // sl_firefox_latest_linux: {
      //   base: `SauceLabs`,
      //   platform: `Linux`,
      //   browserName: `firefox`,
      //   version: `latest`
      // }
      sl_firefox_latest_macOS_High_Sierra: {
        base: 'SauceLabs',
        platform: 'macOS 10.13',
        browserName: 'Firefox',
        version: 'latest',
        extendedDebugging: true,
        'moz:firefoxOptions': {
          args: [
            '-start-debugger-server',
            '9222'
          ],
          prefs: {
            'devtools.chrome.enabled': true,
            'devtools.debugger.prompt-connection': false,
            'devtools.debugger.remote-enabled': true
          }
        }
      }
      // sl_firefox_latest_win7: {
      //   base: `SauceLabs`,
      //   platform: `Windows 7`,
      //   browserName: `firefox`,
      //   version: `latest`
      // }
    };
  }

  return {
    firefox_h264: {
      base: 'FirefoxHeadless'
    },
    ChromeH264: {
      base: 'ChromeHeadless',
      flags: [
        '--use-fake-device-for-media-stream',
        '--use-fake-ui-for-media-stream'
      ]
    }
  };
};
